// @thoughtspot/visual-embed-sdk/cjs/src/utils/processData.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const processDataInstance = tslib_1.__importStar(require("./processData"));
const auth = tslib_1.__importStar(require("../auth"));
const base = tslib_1.__importStar(require("../embed/base"));
const embedConfigInstance = tslib_1.__importStar(require("../embed/embedConfig"));
const types_1 = require("../types");
const sessionInfoService = tslib_1.__importStar(require("./sessionInfoService"));
const utilsModule = tslib_1.__importStar(require("../utils"));
describe('Unit test for process data', () => {
    beforeAll(() => {
        jest.spyOn(auth, 'postLoginService').mockImplementation(() => Promise.resolve(undefined));
        base.init({
            thoughtSpotHost: 'https://tshost',
            authType: types_1.AuthType.None,
        });
    });
    afterEach(() => {
        jest.resetAllMocks();
    });
    const thoughtSpotHost = 'http://localhost';
    test('ProcessData, when Action is CustomAction', async () => {
        const processedData = {
            type: types_1.EmbedEvent.CustomAction,
            data: {},
        };
        jest.spyOn(processDataInstance, 'processCustomAction').mockImplementation(async () => ({}));
        expect(processDataInstance.processEventData(types_1.EmbedEvent.CustomAction, processedData, thoughtSpotHost, null)).toEqual(expect.objectContaining({
            ...processedData,
            answerService: {
                answer: {},
                selectedPoints: undefined,
                session: undefined,
                thoughtSpotHost: 'http://localhost',
                tmlOverride: {},
            },
        }));
    });
    test('ProcessData, when Action is CustomAction with contextMenuPoints', async () => {
        const processedData = {
            type: types_1.EmbedEvent.CustomAction,
            data: {
                contextMenuPoints: {
                    selectedPoints: [{ x: 1, y: 2 }],
                },
            },
        };
        jest.spyOn(processDataInstance, 'processCustomAction').mockImplementation(async () => ({}));
        expect(processDataInstance.processEventData(types_1.EmbedEvent.CustomAction, processedData, thoughtSpotHost, null)).toEqual(expect.objectContaining({
            ...processedData,
            answerService: {
                answer: {},
                selectedPoints: [{ x: 1, y: 2 }],
                session: undefined,
                thoughtSpotHost: 'http://localhost',
                tmlOverride: {},
            },
        }));
    });
    test('ProcessData QueryChanged default branch returns payload unchanged', () => {
        const e = { type: types_1.EmbedEvent.QueryChanged, data: { query: 'revenue by region' } };
        expect(processDataInstance.processEventData(types_1.EmbedEvent.QueryChanged, e, thoughtSpotHost, null)).toEqual(e);
    });
    test('ProcessData, when Action is non CustomAction', () => {
        const processedData = { type: types_1.EmbedEvent.Data };
        jest.spyOn(processDataInstance, 'processCustomAction').mockImplementation(async () => ({}));
        processDataInstance.processEventData(types_1.EmbedEvent.Data, processedData, thoughtSpotHost, null);
        expect(processDataInstance.processCustomAction).not.toHaveBeenCalled();
    });
    test('AuthInit', () => {
        const sessionInfo = {
            userGUID: '1234',
            mixpanelToken: 'abc123',
            isPublicUser: false,
        };
        const e = { type: types_1.EmbedEvent.AuthInit, data: sessionInfo };
        jest.spyOn(base, 'notifyAuthSuccess');
        jest.spyOn(sessionInfoService, 'getSessionInfo').mockImplementation(() => Promise.resolve(sessionInfo));
        expect(processDataInstance.processEventData(e.type, e, '', null)).toEqual({
            type: e.type,
            data: {
                userGUID: sessionInfo.userGUID,
            },
        });
        expect(base.notifyAuthSuccess).toHaveBeenCalled();
    });
    test('NoCookieAccess no suppress alert', () => {
        const e = { type: types_1.EmbedEvent.NoCookieAccess };
        jest.spyOn(base, 'notifyAuthFailure');
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue({
            loginFailedMessage: 'Hello',
            suppressNoCookieAccessAlert: false,
        });
        jest.spyOn(window, 'alert').mockImplementation(() => undefined);
        const el = {};
        expect(processDataInstance.processEventData(e.type, e, '', el)).toEqual({
            type: e.type,
        });
        expect(base.notifyAuthFailure).toHaveBeenCalledWith(auth.AuthFailureType.NO_COOKIE_ACCESS);
        expect(window.alert).toHaveBeenCalled();
        expect(el.innerHTML).toBe('Hello');
    });
    test('NoCookieAccess suppressAlert=true', () => {
        const e = { type: types_1.EmbedEvent.NoCookieAccess };
        jest.spyOn(base, 'notifyAuthFailure');
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue({
            loginFailedMessage: 'Hello',
            suppressNoCookieAccessAlert: true,
        });
        jest.spyOn(window, 'alert').mockReset();
        jest.spyOn(window, 'alert').mockImplementation(() => undefined);
        const el = {};
        expect(processDataInstance.processEventData(e.type, e, '', el)).toEqual({
            type: e.type,
        });
        expect(base.notifyAuthFailure).toHaveBeenCalledWith(auth.AuthFailureType.NO_COOKIE_ACCESS);
        expect(window.alert).not.toHaveBeenCalled();
        expect(el.innerHTML).toBe('Hello');
    });
    test('NoCookieAccess ignoreNoCookieAccess=true', () => {
        const e = { type: types_1.EmbedEvent.NoCookieAccess };
        jest.spyOn(base, 'notifyAuthFailure');
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue({
            loginFailedMessage: 'Hello',
            ignoreNoCookieAccess: true,
        });
        jest.spyOn(window, 'alert').mockReset();
        jest.spyOn(window, 'alert').mockImplementation(() => undefined);
        const el = {};
        expect(processDataInstance.processEventData(e.type, e, '', el)).toEqual({
            type: e.type,
        });
        expect(base.notifyAuthFailure).toHaveBeenCalledWith(auth.AuthFailureType.NO_COOKIE_ACCESS);
        expect(window.alert).not.toHaveBeenCalled();
        expect(el.innerHTML).not.toBe('Hello');
    });
    test('process authFailure', () => {
        const e = { type: types_1.EmbedEvent.AuthFailure };
        jest.spyOn(base, 'notifyAuthFailure');
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue({
            loginFailedMessage: 'Hello',
        });
        const el = {};
        expect(processDataInstance.processEventData(e.type, e, '', el)).toEqual({
            type: e.type,
        });
        expect(base.notifyAuthFailure).toHaveBeenCalledWith(auth.AuthFailureType.OTHER);
        expect(el.innerHTML).toBe('Hello');
    });
    test('process authFailure AuthType=None', () => {
        const e = { type: types_1.EmbedEvent.AuthFailure };
        jest.spyOn(base, 'notifyAuthFailure');
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue({
            loginFailedMessage: 'Hello',
            authType: types_1.AuthType.None,
        });
        const el = {};
        expect(processDataInstance.processEventData(e.type, e, '', el)).toEqual({
            type: e.type,
        });
        expect(base.notifyAuthFailure).not.toHaveBeenCalled();
        expect(el.innerHTML).not.toBe('Hello');
    });
    test('process authLogout', () => {
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockRestore();
        base.init({
            loginFailedMessage: 'Hello',
            autoLogin: true,
            thoughtSpotHost: 'https://tshost',
            authType: types_1.AuthType.None,
        });
        const e = { type: types_1.EmbedEvent.AuthLogout };
        jest.spyOn(base, 'notifyLogout');
        const el = {};
        expect(processDataInstance.processEventData(e.type, e, '', el)).toEqual({
            type: e.type,
        });
        expect(base.notifyLogout).toHaveBeenCalled();
        expect(el.innerHTML).toBe('Hello');
        expect(embedConfigInstance.getEmbedConfig().autoLogin).toBe(false);
    });
    test('process authFailure AuthType=None', () => {
        const e = { type: types_1.EmbedEvent.AuthFailure };
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue({
            loginFailedMessage: 'Hello',
            authType: types_1.AuthType.EmbeddedSSO,
            disableLoginFailurePage: true,
        });
        const el = {};
        expect(processDataInstance.processEventData(e.type, e, '', el)).toEqual({
            type: e.type,
        });
        expect(base.notifyAuthFailure).not.toHaveBeenCalled();
        expect(el.innerHTML).not.toBe('Hello');
    });
    test('process authFailure AuthType=TrustedAuthToken and autoLogin true', () => {
        const e = { type: types_1.EmbedEvent.AuthFailure };
        jest.spyOn(base, 'notifyAuthFailure');
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue({
            loginFailedMessage: 'Hello',
            authType: types_1.AuthType.TrustedAuthToken,
            autoLogin: true,
        });
        const el = {};
        expect(processDataInstance.processEventData(e.type, e, '', el)).toEqual({
            type: e.type,
        });
        expect(base.notifyAuthFailure).toHaveBeenCalledWith(auth.AuthFailureType.IDLE_SESSION_TIMEOUT);
        expect(el.innerHTML).toBe('Hello');
    });
    test('process authFailure with TrustedAuthTokenCookieless and autoLogin', () => {
        const e = { type: types_1.EmbedEvent.AuthFailure };
        jest.spyOn(base, 'notifyAuthFailure');
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue({
            loginFailedMessage: 'Hello',
            authType: types_1.AuthType.TrustedAuthTokenCookieless,
            autoLogin: true,
        });
        const el = {};
        expect(processDataInstance.processEventData(e.type, e, '', el)).toEqual({
            type: e.type,
        });
        expect(base.notifyAuthFailure).toHaveBeenCalledWith(auth.AuthFailureType.IDLE_SESSION_TIMEOUT);
        expect(el.innerHTML).toBe('Hello');
    });
    test('should handle ExitPresentMode when disableFullscreenPresentation is false (enabled)', () => {
        const mockConfig = {
            disableFullscreenPresentation: false,
        };
        const mockHandleExitPresentMode = jest.spyOn(utilsModule, 'handleExitPresentMode').mockImplementation(() => Promise.resolve(undefined));
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue(mockConfig);
        const processedData = {
            type: types_1.EmbedEvent.ExitPresentMode,
            data: {},
        };
        processDataInstance.processEventData(types_1.EmbedEvent.ExitPresentMode, processedData, thoughtSpotHost, null);
        expect(mockHandleExitPresentMode).toHaveBeenCalled();
        mockHandleExitPresentMode.mockReset();
    });
    test('AuthInit with payload.userGUID fallback when data.userGUID is absent', () => {
        const e = {
            type: types_1.EmbedEvent.AuthInit,
            payload: { userGUID: 'payloadGUID' },
        };
        jest.spyOn(base, 'notifyAuthSuccess');
        expect(processDataInstance.processEventData(e.type, e, '', null)).toEqual({
            type: e.type,
            payload: { userGUID: 'payloadGUID' },
            data: { userGUID: 'payloadGUID' },
        });
        expect(base.notifyAuthSuccess).toHaveBeenCalled();
    });
    test('process authFailure EmbeddedSSO with UNAUTHENTICATED_FAILURE suppresses login page', () => {
        const e = {
            type: types_1.EmbedEvent.AuthFailure,
            data: { type: auth.AuthFailureType.UNAUTHENTICATED_FAILURE },
        };
        jest.spyOn(base, 'notifyAuthFailure');
        jest.spyOn(embedConfigInstance, 'getEmbedConfig').mockReturnValue({
            loginFailedMessage: 'Hello',
            authType: types_1.AuthType.EmbeddedSSO,
            disableLoginFailurePage: false,
        });
        const el = {};
        processDataInstance.processEventData(e.type, e, '', el);
        // isEmbeddedSSOInfoFailure=true so neither branch fires, innerHTML stays unset
        expect(el.innerHTML).toBeUndefined();
        expect(base.notifyAuthFailure).not.toHaveBeenCalled();
    });
    test('processEventData default branch returns payload unchanged for non-special events', () => {
        const events = [
            types_1.EmbedEvent.Drilldown,
            types_1.EmbedEvent.DataSourceSelected,
            types_1.EmbedEvent.AddRemoveColumns,
            types_1.EmbedEvent.VizPointDoubleClick,
            types_1.EmbedEvent.VizPointClick,
            types_1.EmbedEvent.Alert,
            types_1.EmbedEvent.GetDataClick,
            types_1.EmbedEvent.DialogClose,
            types_1.EmbedEvent.Download,
            types_1.EmbedEvent.DownloadAsPng,
            types_1.EmbedEvent.DownloadAsPdf,
            types_1.EmbedEvent.DownloadAsCsv,
            types_1.EmbedEvent.DownloadAsXlsx,
            types_1.EmbedEvent.DownloadLiveboardAsContinuousPDF,
            types_1.EmbedEvent.AnswerDelete,
            types_1.EmbedEvent.AIHighlights,
            types_1.EmbedEvent.Pin,
            types_1.EmbedEvent.SpotIQAnalyze,
            types_1.EmbedEvent.Share,
            types_1.EmbedEvent.DrillInclude,
            types_1.EmbedEvent.DrillExclude,
            types_1.EmbedEvent.CopyToClipboard,
            types_1.EmbedEvent.UpdateTML,
            types_1.EmbedEvent.EditTML,
            types_1.EmbedEvent.ExportTML,
            types_1.EmbedEvent.SaveAsView,
            types_1.EmbedEvent.CopyAEdit,
            types_1.EmbedEvent.ShowUnderlyingData,
            types_1.EmbedEvent.AnswerChartSwitcher,
            types_1.EmbedEvent.LiveboardInfo,
            types_1.EmbedEvent.AddToFavorites,
            types_1.EmbedEvent.Schedule,
            types_1.EmbedEvent.Edit,
            types_1.EmbedEvent.MakeACopy,
            types_1.EmbedEvent.Present,
            types_1.EmbedEvent.Delete,
            types_1.EmbedEvent.SchedulesList,
            types_1.EmbedEvent.Cancel,
            types_1.EmbedEvent.Explore,
            types_1.EmbedEvent.CopyLink,
            types_1.EmbedEvent.CrossFilterChanged,
            types_1.EmbedEvent.VizPointRightClick,
            types_1.EmbedEvent.InsertIntoSlide,
            types_1.EmbedEvent.FilterChanged,
            types_1.EmbedEvent.UpdateConnection,
            types_1.EmbedEvent.CreateConnection,
            types_1.EmbedEvent.ResetLiveboard,
            types_1.EmbedEvent.ChangePersonalizedView,
            types_1.EmbedEvent.CreateWorksheet,
            types_1.EmbedEvent.AskSageInit,
            types_1.EmbedEvent.Rename,
            types_1.EmbedEvent.ParameterChanged,
            types_1.EmbedEvent.TableVizRendered,
            types_1.EmbedEvent.CreateLiveboard,
            types_1.EmbedEvent.CreateModel,
            types_1.EmbedEvent.SpotterData,
            types_1.EmbedEvent.PreviewSpotterData,
            types_1.EmbedEvent.AddToCoaching,
            types_1.EmbedEvent.DataModelInstructions,
            types_1.EmbedEvent.SpotterQueryTriggered,
            types_1.EmbedEvent.LastPromptEdited,
            types_1.EmbedEvent.LastPromptDeleted,
            types_1.EmbedEvent.ResetSpotterConversation,
            types_1.EmbedEvent.SpotterInit,
            types_1.EmbedEvent.SpotterLoadComplete,
            types_1.EmbedEvent.OrgSwitched,
            types_1.EmbedEvent.SpotterConversationRenamed,
            types_1.EmbedEvent.SpotterConversationDeleted,
            types_1.EmbedEvent.SpotterConversationSelected,
            types_1.EmbedEvent.EmbedPageContextChanged,
            types_1.EmbedEvent.Subscribed,
            types_1.EmbedEvent.SendTestScheduleEmail,
            types_1.EmbedEvent.SpotterVizInit,
            types_1.EmbedEvent.SpotterVizQueryTriggered,
            types_1.EmbedEvent.SpotterVizResponseComplete,
            types_1.EmbedEvent.SpotterVizCheckpointCreated,
            types_1.EmbedEvent.SpotterVizCheckpointRestored,
            types_1.EmbedEvent.SpotterVizError,
            types_1.EmbedEvent.SpotterVizClosed,
            types_1.EmbedEvent.RefreshLiveboardBrowserCache,
            types_1.EmbedEvent.V1Data,
        ];
        events.forEach((eventType) => {
            const e = { type: eventType, data: 'test-payload' };
            expect(processDataInstance.processEventData(eventType, e, thoughtSpotHost, null)).toEqual(e);
        });
    });
    test('should handle ClearInfoCache', () => {
        const mockResetCachedPreauthInfo = jest.spyOn(sessionInfoService, 'resetCachedPreauthInfo').mockImplementation(() => { });
        const mockResetCachedSessionInfo = jest.spyOn(sessionInfoService, 'resetCachedSessionInfo').mockImplementation(() => { });
        const processedData = {
            type: types_1.EmbedEvent.CLEAR_INFO_CACHE,
            data: {},
        };
        processDataInstance.processEventData(types_1.EmbedEvent.CLEAR_INFO_CACHE, processedData, thoughtSpotHost, null);
        expect(mockResetCachedPreauthInfo).toHaveBeenCalled();
        expect(mockResetCachedSessionInfo).toHaveBeenCalled();
    });
});
