// @thoughtspot/visual-embed-sdk/cjs/src/embed/app.js
"use strict";
/**
 * Copyright (c) 2022
 *
 * Full application embedding
 * https://developers.thoughtspot.com/docs/?pageid=full-embed
 * @summary Full app embed
 * @module
 * @author Ayon Ghosh <ayon.ghosh@thoughtspot.com>
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppEmbed = exports.ListPage = exports.HomePage = exports.PrimaryNavbarVersion = exports.HomePageSearchBarMode = exports.DataPanelCustomColumnGroupsAccordionState = exports.Page = void 0;
const logger_1 = require("../utils/logger");
const utils_1 = require("../utils");
const types_1 = require("../types");
const ts_embed_1 = require("./ts-embed");
const spotter_utils_1 = require("./spotter-utils");
const spotter_viz_utils_1 = require("./spotter-viz-utils");
/**
 * Pages within the ThoughtSpot app that can be embedded.
 */
var Page;
(function (Page) {
    /**
     * Home page
     */
    Page["Home"] = "home";
    /**
     * Search page
     */
    Page["Search"] = "search";
    /**
     * Saved answers listing page
     */
    Page["Answers"] = "answers";
    /**
     * Liveboards listing page
     */
    Page["Liveboards"] = "liveboards";
    /**
     * @hidden
     */
    Page["Pinboards"] = "pinboards";
    /**
     * Data management page
     */
    Page["Data"] = "data";
    /**
     * SpotIQ listing page
     */
    Page["SpotIQ"] = "insights";
    /**
     *  Monitor Alerts Page
     */
    Page["Monitor"] = "monitor";
    /**
     * Collections listing page
     */
    Page["Collections"] = "collections";
})(Page = exports.Page || (exports.Page = {}));
/**
 * Define the initial state of column custom group accordions
 * in data panel v2.
 */
var DataPanelCustomColumnGroupsAccordionState;
(function (DataPanelCustomColumnGroupsAccordionState) {
    /**
     * Expand all the accordion initially in data panel v2.
     */
    DataPanelCustomColumnGroupsAccordionState["EXPAND_ALL"] = "EXPAND_ALL";
    /**
     * Collapse all the accordions initially in data panel v2.
     */
    DataPanelCustomColumnGroupsAccordionState["COLLAPSE_ALL"] = "COLLAPSE_ALL";
    /**
     * Expand the first accordion and collapse the rest.
     */
    DataPanelCustomColumnGroupsAccordionState["EXPAND_FIRST"] = "EXPAND_FIRST";
})(DataPanelCustomColumnGroupsAccordionState = exports.DataPanelCustomColumnGroupsAccordionState || (exports.DataPanelCustomColumnGroupsAccordionState = {}));
var HomePageSearchBarMode;
(function (HomePageSearchBarMode) {
    HomePageSearchBarMode["OBJECT_SEARCH"] = "objectSearch";
    HomePageSearchBarMode["AI_ANSWER"] = "aiAnswer";
    HomePageSearchBarMode["NONE"] = "none";
})(HomePageSearchBarMode = exports.HomePageSearchBarMode || (exports.HomePageSearchBarMode = {}));
/**
 * Define the version of the primary navbar
 * @version SDK: 1.40.0 | ThoughtSpot: 10.11.0.cl
 */
var PrimaryNavbarVersion;
(function (PrimaryNavbarVersion) {
    /**
     * Sliding (v3) introduces a new left-side navigation hub featuring a tab switcher,
     * along with updates to the top navigation bar.
     * It serves as the foundational version of the PrimaryNavBar.
     */
    PrimaryNavbarVersion["Sliding"] = "v3";
})(PrimaryNavbarVersion = exports.PrimaryNavbarVersion || (exports.PrimaryNavbarVersion = {}));
/**
 * Define the version of the home page
 * @version SDK: 1.40.0 | ThoughtSpot: 10.11.0.cl
 */
var HomePage;
(function (HomePage) {
    /**
     * Modular (v2) introduces the updated Modular Home Experience.
     * It serves as the foundational version of the home page.
     */
    HomePage["Modular"] = "v2";
    /**
     * ModularWithStylingChanges (v3) introduces Modular Home Experience
     * with styling changes.
     */
    HomePage["ModularWithStylingChanges"] = "v3";
    /**
     * Focused (v4) introduces the V4 homepage experience
     * in which Watchlist and recents and incorporated together
     * to form a more focused homepage.
     * Pre-requisite : spotter enablement
     * @version SDK: 1.50.0 | ThoughtSpot Cloud: 26.7.0.cl
     */
    HomePage["Focused"] = "v4";
})(HomePage = exports.HomePage || (exports.HomePage = {}));
/**
 * Define the version of the list page
 * @version SDK: 1.40.0 | ThoughtSpot: 10.12.0.cl
 */
var ListPage;
(function (ListPage) {
    /**
     * List (v2) is the traditional List Experience.
     * It serves as the foundational version of the list page.
     */
    ListPage["List"] = "v2";
    /**
     * ListWithUXChanges (v3) introduces the new updated list page with UX changes.
     */
    ListPage["ListWithUXChanges"] = "v3";
})(ListPage = exports.ListPage || (exports.ListPage = {}));
/**
 * Embeds full ThoughtSpot experience in a host application.
 * @group Embed components
 */
class AppEmbed extends ts_embed_1.V1Embed {
    constructor(domSelector, viewConfig) {
        viewConfig.embedComponentType = 'AppEmbed';
        super(domSelector, viewConfig);
        this.defaultHeight = 500;
        this.lazyLoadScrollContainers = [];
        this.sendFullHeightLazyLoadData = () => {
            const data = (0, utils_1.calculateVisibleElementData)(this.iFrame, this.viewConfig.enableScrollableContainerLazyLoading);
            // this should be fired only if the lazyLoadingForFullHeight and fullHeight are true
            if (this.viewConfig.lazyLoadingForFullHeight && this.viewConfig.fullHeight) {
                this.trigger(types_1.HostEvent.VisibleEmbedCoordinates, data);
            }
        };
        /**
         * This is a handler for the RequestVisibleEmbedCoordinates event.
         * It is used to send the visible coordinates data to the host application.
         * @param data The event payload
         * @param responder The responder function
         */
        this.requestVisibleEmbedCoordinatesHandler = (data, responder) => {
            logger_1.logger.info('Sending RequestVisibleEmbedCoordinates', data);
            const visibleCoordinatesData = (0, utils_1.calculateVisibleElementData)(this.iFrame, this.viewConfig.enableScrollableContainerLazyLoading);
            responder({ type: types_1.EmbedEvent.RequestVisibleEmbedCoordinates, data: visibleCoordinatesData });
        };
        /**
         * Set the iframe height as per the computed height received
         * from the ThoughtSpot app.
         * @param data The event payload
         */
        this.updateIFrameHeight = (data) => {
            this.setIFrameHeight(Math.max(data.data, this.defaultHeight));
            this.sendFullHeightLazyLoadData();
        };
        this.embedIframeCenter = (data, responder) => {
            const obj = this.getIframeCenter();
            responder({ type: types_1.EmbedEvent.EmbedIframeCenter, data: obj });
        };
        this.setIframeHeightForNonEmbedLiveboard = (data) => {
            const { height: frameHeight } = this.viewConfig.frameParams || {};
            const liveboardRelatedRoutes = [
                '/pinboard/',
                '/insights/pinboard/',
                '/schedules/',
                '/embed/viz/',
                '/embed/insights/viz/',
                '/liveboard/',
                '/insights/liveboard/',
                '/tsl-editor/PINBOARD_ANSWER_BOOK/',
                '/import-tsl/PINBOARD_ANSWER_BOOK/',
            ];
            if (liveboardRelatedRoutes.some((path) => data.data.currentPath.startsWith(path))) {
                // Ignore the height reset of the frame, if the navigation is
                // only within the liveboard page.
                return;
            }
            this.setIFrameHeight(frameHeight || this.defaultHeight);
        };
        if (this.viewConfig.fullHeight === true) {
            this.on(types_1.EmbedEvent.RouteChange, this.setIframeHeightForNonEmbedLiveboard);
            this.on(types_1.EmbedEvent.EmbedHeight, this.updateIFrameHeight);
            this.on(types_1.EmbedEvent.EmbedIframeCenter, this.embedIframeCenter);
            this.on(types_1.EmbedEvent.RequestVisibleEmbedCoordinates, this.requestVisibleEmbedCoordinatesHandler);
        }
    }
    /**
     * Extends the default APP_INIT payload with `embedParams.spotterSidebarConfig`
     * so the conv-assist app can read sidebar configuration on initialisation.
     *
     * Precedence for `enablePastConversationsSidebar`:
     * `spotterSidebarConfig.enablePastConversationsSidebar` wins over the
     * deprecated top-level `enablePastConversationsSidebar` flag; if the former
     * is absent the latter is used as a fallback.
     *
     * An invalid `spotterDocumentationUrl` triggers a validation error and is
     * excluded from the payload rather than forwarded to the app.
     */
    async getAppInitData() {
        const defaultAppInitData = await super.getAppInitData();
        const sidebarInitData = (0, spotter_utils_1.buildSpotterSidebarAppInitData)(defaultAppInitData, this.viewConfig, this.handleError.bind(this));
        return (0, spotter_viz_utils_1.buildSpotterVizAppInitData)(sidebarInitData, this.viewConfig);
    }
    /**
     * Constructs a map of parameters to be passed on to the
     * embedded Liveboard or visualization.
     */
    getEmbedParams() {
        const { tag, hideTagFilterChips, hideObjects, liveboardV2, showPrimaryNavbar, disableProfileAndHelp, hideHamburger, hideObjectSearch, hideNotification, hideApplicationSwitcher, hideOrgSwitcher, enableSearchAssist, newConnectionsExperience, fullHeight, dataPanelV2 = true, hideLiveboardHeader = false, showLiveboardTitle = true, showLiveboardDescription = true, showMaskedFilterChip = false, isLiveboardMasterpiecesEnabled = false, newChartsLibrary, hideHomepageLeftNav = false, modularHomeExperience = false, isLiveboardHeaderSticky = true, enableAskSage, collapseSearchBarInitially = false, enable2ColumnLayout, enableCustomColumnGroups = false, dataPanelCustomGroupsAccordionInitialState = DataPanelCustomColumnGroupsAccordionState.EXPAND_ALL, collapseSearchBar = true, isLiveboardCompactHeaderEnabled = false, showLiveboardVerifiedBadge = true, showLiveboardReverifyBanner = true, hideIrrelevantChipsInLiveboardTabs = false, isEnhancedFilterInteractivityEnabled = false, homePageSearchBarMode, isUnifiedSearchExperienceEnabled = true, enablePendoHelp = true, discoveryExperience, coverAndFilterOptionInPDF = false, isLiveboardStylingAndGroupingEnabled, isPNGInScheduledEmailsEnabled = false, isLiveboardXLSXCSVDownloadEnabled = false, isGranularXLSXCSVSchedulesEnabled = false, isCentralizedLiveboardFilterUXEnabled = false, isLinkParametersEnabled, updatedSpotterChatPrompt, defaultQueryMode, enableStopAnswerGenerationEmbed, spotterChatConfig, minimumHeight, isThisPeriodInDateFiltersEnabled, enableHomepageAnnouncement = false, isContinuousLiveboardPDFEnabled = false, enableLiveboardDataCache, } = this.viewConfig;
        let params = {};
        params[types_1.Param.PrimaryNavHidden] = !showPrimaryNavbar;
        params[types_1.Param.HideProfleAndHelp] = !!disableProfileAndHelp;
        params[types_1.Param.HideApplicationSwitcher] = !!hideApplicationSwitcher;
        params[types_1.Param.HideOrgSwitcher] = !!hideOrgSwitcher;
        params[types_1.Param.HideLiveboardHeader] = hideLiveboardHeader;
        params[types_1.Param.ShowLiveboardTitle] = showLiveboardTitle;
        params[types_1.Param.ShowLiveboardDescription] = !!showLiveboardDescription;
        params[types_1.Param.ShowMaskedFilterChip] = showMaskedFilterChip;
        params[types_1.Param.IsLiveboardMasterpiecesEnabled] = isLiveboardMasterpiecesEnabled;
        if (newChartsLibrary !== undefined) {
            params[types_1.Param.EnableNewChartLibrary] = newChartsLibrary;
        }
        params[types_1.Param.LiveboardHeaderSticky] = isLiveboardHeaderSticky;
        params[types_1.Param.IsFullAppEmbed] = true;
        params[types_1.Param.LiveboardHeaderV2] = isLiveboardCompactHeaderEnabled;
        params[types_1.Param.IsEnhancedFilterInteractivityEnabled] = isEnhancedFilterInteractivityEnabled;
        params[types_1.Param.ShowLiveboardVerifiedBadge] = showLiveboardVerifiedBadge;
        params[types_1.Param.ShowLiveboardReverifyBanner] = showLiveboardReverifyBanner;
        params[types_1.Param.HideIrrelevantFiltersInTab] = hideIrrelevantChipsInLiveboardTabs;
        if (isUnifiedSearchExperienceEnabled !== undefined) {
            params[types_1.Param.IsUnifiedSearchExperienceEnabled] = isUnifiedSearchExperienceEnabled;
        }
        params[types_1.Param.CoverAndFilterOptionInPDF] = !!coverAndFilterOptionInPDF;
        params = this.getBaseQueryParams(params);
        if (!(0, utils_1.isUndefined)(updatedSpotterChatPrompt)) {
            params[types_1.Param.UpdatedSpotterChatPrompt] = !!updatedSpotterChatPrompt;
        }
        if (!(0, utils_1.isUndefined)(defaultQueryMode)) {
            params[types_1.Param.DefaultQueryMode] = defaultQueryMode;
        }
        if (!(0, utils_1.isUndefined)(enableStopAnswerGenerationEmbed)) {
            params[types_1.Param.EnableStopAnswerGenerationEmbed] = !!enableStopAnswerGenerationEmbed;
        }
        // Handle spotterChatConfig params
        if (spotterChatConfig) {
            const { hideToolResponseCardBranding, toolResponseCardBrandingLabel, spotterFileUploadEnabled, spotterFileUploadFileTypes, enableStarterPrompts, } = spotterChatConfig;
            (0, utils_1.setParamIfDefined)(params, types_1.Param.HideToolResponseCardBranding, hideToolResponseCardBranding, true);
            (0, utils_1.setParamIfDefined)(params, types_1.Param.ToolResponseCardBrandingLabel, toolResponseCardBrandingLabel);
            if (spotterFileUploadEnabled !== undefined) {
                params[types_1.Param.SpotterFileUploadEnabled] = spotterFileUploadEnabled;
            }
            if (enableStarterPrompts !== undefined) {
                params[types_1.Param.IsStarterPromptsEnabled] = enableStarterPrompts;
            }
            if (spotterFileUploadFileTypes !== undefined) {
                params[types_1.Param.SpotterFileUploadFileTypes] = JSON.stringify(spotterFileUploadFileTypes);
            }
        }
        if (hideObjectSearch) {
            params[types_1.Param.HideObjectSearch] = !!hideObjectSearch;
        }
        if (hideHamburger) {
            params[types_1.Param.HideHamburger] = !!hideHamburger;
        }
        if (hideNotification) {
            params[types_1.Param.HideNotification] = !!hideNotification;
        }
        if (fullHeight === true) {
            params[types_1.Param.fullHeight] = true;
            if (this.viewConfig.lazyLoadingForFullHeight) {
                params[types_1.Param.IsLazyLoadingForEmbedEnabled] = true;
                if ((0, utils_1.isValidCssMargin)(this.viewConfig.lazyLoadingMargin)) {
                    params[types_1.Param.RootMarginForLazyLoad] = this.viewConfig.lazyLoadingMargin;
                }
            }
        }
        if (tag) {
            params[types_1.Param.Tag] = tag;
        }
        if (hideObjects && hideObjects.length) {
            params[types_1.Param.HideObjects] = JSON.stringify(hideObjects);
        }
        if (liveboardV2 !== undefined) {
            params[types_1.Param.LiveboardV2Enabled] = liveboardV2;
        }
        if (enableSearchAssist !== undefined) {
            params[types_1.Param.EnableSearchAssist] = enableSearchAssist;
        }
        if (newConnectionsExperience !== undefined) {
            params[types_1.Param.EnableConnectionNewExperience] = newConnectionsExperience;
        }
        if (enable2ColumnLayout !== undefined) {
            params[types_1.Param.Enable2ColumnLayout] = enable2ColumnLayout;
        }
        if (enableAskSage) {
            params[types_1.Param.enableAskSage] = enableAskSage;
        }
        if (homePageSearchBarMode) {
            params[types_1.Param.HomePageSearchBarMode] = homePageSearchBarMode;
        }
        if (enablePendoHelp !== undefined) {
            params[types_1.Param.EnablePendoHelp] = enablePendoHelp;
        }
        if (isLiveboardStylingAndGroupingEnabled !== undefined) {
            params[types_1.Param.IsLiveboardStylingAndGroupingEnabled] = isLiveboardStylingAndGroupingEnabled;
        }
        if (isPNGInScheduledEmailsEnabled !== undefined) {
            params[types_1.Param.isPNGInScheduledEmailsEnabled] = isPNGInScheduledEmailsEnabled;
        }
        if (isLiveboardXLSXCSVDownloadEnabled !== undefined) {
            params[types_1.Param.isLiveboardXLSXCSVDownloadEnabled] = isLiveboardXLSXCSVDownloadEnabled;
        }
        if (isGranularXLSXCSVSchedulesEnabled !== undefined) {
            params[types_1.Param.isGranularXLSXCSVSchedulesEnabled] = isGranularXLSXCSVSchedulesEnabled;
        }
        if (hideTagFilterChips !== undefined) {
            params[types_1.Param.HideTagFilterChips] = hideTagFilterChips;
        }
        if (isLinkParametersEnabled !== undefined) {
            params[types_1.Param.isLinkParametersEnabled] = isLinkParametersEnabled;
        }
        if (isCentralizedLiveboardFilterUXEnabled != undefined) {
            params[types_1.Param.isCentralizedLiveboardFilterUXEnabled] = isCentralizedLiveboardFilterUXEnabled;
        }
        if (isThisPeriodInDateFiltersEnabled !== undefined) {
            params[types_1.Param.IsThisPeriodInDateFiltersEnabled] = isThisPeriodInDateFiltersEnabled;
        }
        if (enableHomepageAnnouncement !== undefined) {
            params[types_1.Param.EnableHomepageAnnouncement] = enableHomepageAnnouncement;
        }
        if (isContinuousLiveboardPDFEnabled !== undefined) {
            params[types_1.Param.IsWYSIWYGLiveboardPDFEnabled] = isContinuousLiveboardPDFEnabled;
        }
        this.defaultHeight = minimumHeight || this.defaultHeight;
        if (enableLiveboardDataCache !== undefined) {
            params[types_1.Param.EnableLiveboardDataCache] = enableLiveboardDataCache;
        }
        params[types_1.Param.DataPanelV2Enabled] = dataPanelV2;
        params[types_1.Param.HideHomepageLeftNav] = hideHomepageLeftNav;
        params[types_1.Param.ModularHomeExperienceEnabled] = modularHomeExperience;
        params[types_1.Param.CollapseSearchBarInitially] = collapseSearchBarInitially || collapseSearchBar;
        params[types_1.Param.EnableCustomColumnGroups] = enableCustomColumnGroups;
        if (dataPanelCustomGroupsAccordionInitialState
            === DataPanelCustomColumnGroupsAccordionState.COLLAPSE_ALL
            || dataPanelCustomGroupsAccordionInitialState
                === DataPanelCustomColumnGroupsAccordionState.EXPAND_FIRST) {
            params[types_1.Param.DataPanelCustomGroupsAccordionInitialState] = dataPanelCustomGroupsAccordionInitialState;
        }
        else {
            params[types_1.Param.DataPanelCustomGroupsAccordionInitialState] = DataPanelCustomColumnGroupsAccordionState.EXPAND_ALL;
        }
        if (modularHomeExperience !== undefined) {
            params[types_1.Param.ModularHomeExperienceEnabled] = modularHomeExperience;
        }
        // Set navigation to v2 by default to avoid problems like the app
        // switcher (9-dot menu) not showing when v3 navigation is turned on
        // at the cluster level.
        // To use v3 navigation, we must manually set the discoveryExperience
        // settings.
        params[types_1.Param.NavigationVersion] = 'v2';
        // Set homePageVersion to v2 by default to reset the LD flag value
        // for the homepageVersion.
        params[types_1.Param.HomepageVersion] = 'v2';
        if (discoveryExperience) {
            // primaryNavbarVersion v3 will enabled the new left navigation
            if (discoveryExperience.primaryNavbarVersion === PrimaryNavbarVersion.Sliding) {
                params[types_1.Param.NavigationVersion] = discoveryExperience.primaryNavbarVersion;
                // Enable the modularHomeExperience when Sliding is enabled.
                params[types_1.Param.ModularHomeExperienceEnabled] = true;
            }
            // homePage v2 will enable the modular home page
            // and it will override the modularHomeExperience value
            if (discoveryExperience.homePage === HomePage.Modular) {
                params[types_1.Param.ModularHomeExperienceEnabled] = true;
            }
            // ModularWithStylingChanges (v3) introduces the styling changes
            // to the Modular Homepage.
            // v3 will be the base version of homePageVersion.
            if (discoveryExperience.homePage === HomePage.ModularWithStylingChanges) {
                params[types_1.Param.HomepageVersion] = HomePage.ModularWithStylingChanges;
            }
            // listPageVersion can be changed to v2 or v3
            if (discoveryExperience.listPageVersion !== undefined) {
                params[types_1.Param.ListPageVersion] = discoveryExperience.listPageVersion;
            }
            if (discoveryExperience.homePage === HomePage.Focused) {
                params[types_1.Param.HomepageVersion] = HomePage.Focused;
                // The Focused (V4) homepage experience requires the updated
                // Spotter chat prompt. Enable it automatically unless the
                // developer has explicitly set updatedSpotterChatPrompt.
                if ((0, utils_1.isUndefined)(updatedSpotterChatPrompt)) {
                    params[types_1.Param.UpdatedSpotterChatPrompt] = true;
                }
            }
        }
        const queryParams = (0, utils_1.getQueryParamString)(params, true);
        return queryParams;
    }
    /**
     * Constructs the URL of the ThoughtSpot app page to be rendered.
     * @param pageId The ID of the page to be embedded.
     */
    getIFrameSrc() {
        const { pageId, path, modularHomeExperience } = this.viewConfig;
        const pageRoute = this.formatPath(path) || this.getPageRoute(pageId, modularHomeExperience);
        let url = `${this.getRootIframeSrc()}/${pageRoute}`;
        const tsPostHashParams = this.getThoughtSpotPostUrlParams();
        url = `${url}${tsPostHashParams}`;
        return url;
    }
    /**
     * Gets the ThoughtSpot route of the page for a particular page ID.
     * @param pageId The identifier for a page in the ThoughtSpot app.
     * @param modularHomeExperience
     */
    getPageRoute(pageId, modularHomeExperience = false) {
        switch (pageId) {
            case Page.Search:
                return 'answer';
            case Page.Answers:
                return modularHomeExperience ? 'home/answers' : 'answers';
            case Page.Liveboards:
                return modularHomeExperience ? 'home/liveboards' : 'pinboards';
            case Page.Pinboards:
                return modularHomeExperience ? 'home/liveboards' : 'pinboards';
            case Page.Data:
                return 'data/tables';
            case Page.SpotIQ:
                return modularHomeExperience ? 'home/spotiq-analysis' : 'insights/results';
            case Page.Monitor:
                return modularHomeExperience ? 'home/monitor-alerts' : 'insights/monitor-alerts';
            case Page.Collections:
                return 'collections';
            case Page.Home:
            default:
                return 'home';
        }
    }
    /**
     * Formats the path provided by the user.
     * @param path The URL path.
     * @returns The URL path that the embedded app understands.
     */
    formatPath(path) {
        if (!path) {
            return null;
        }
        // remove leading slash
        if (path.indexOf('/') === 0) {
            return path.substring(1);
        }
        return path;
    }
    /**
     * Navigate to particular page for app embed. eg:answers/pinboards/home
     * This is used for embedding answers, pinboards, visualizations and full application
     * only.
     * @param path string | number The string, set to iframe src and navigate to new page
     * eg: appEmbed.navigateToPage('pinboards')
     * When used with `noReload` (default: true) this can also be a number
     * like 1/-1 to go forward/back.
     * @param noReload boolean Trigger the navigation without reloading the page
     * @version SDK: 1.12.0 | ThoughtSpot: 8.4.0.cl, 8.4.1-sw
     */
    navigateToPage(path, noReload = false) {
        var _a;
        if (!this.iFrame) {
            logger_1.logger.log('Please call render before invoking this method');
            return;
        }
        const overrideHistoryState = (_a = this.viewConfig) === null || _a === void 0 ? void 0 : _a.overrideHistoryState;
        if (noReload || overrideHistoryState) {
            this.trigger(types_1.HostEvent.Navigate, path);
        }
        else {
            if (typeof path !== 'string') {
                logger_1.logger.warn('Path can only by a string when triggered without noReload');
                return;
            }
            const iframeSrc = this.iFrame.src;
            const embedPath = '#/embed';
            const currentPath = iframeSrc.includes(embedPath) ? embedPath : '#';
            this.iFrame.src = `${iframeSrc.split(currentPath)[0]}${currentPath}/${path.replace(/^\/?#?\//, '')}`;
        }
    }
    /**
     * Destroys the ThoughtSpot embed, and remove any nodes from the DOM.
     * @version SDK: 1.39.0 | ThoughtSpot: 10.10.0.cl
     */
    destroy() {
        super.destroy();
        this.unregisterLazyLoadEvents();
    }
    postRender() {
        this.registerLazyLoadEvents();
    }
    registerLazyLoadEvents() {
        if (!this.iFrame) {
            return;
        }
        if (this.viewConfig.fullHeight && this.viewConfig.lazyLoadingForFullHeight) {
            this.unregisterLazyLoadEvents();
            // TODO: Use passive: true, install modernizr to check for passive
            window.addEventListener('resize', this.sendFullHeightLazyLoadData);
            window.addEventListener('scroll', this.sendFullHeightLazyLoadData, true);
            if (!this.viewConfig.enableScrollableContainerLazyLoading) {
                return;
            }
            this.lazyLoadScrollContainers = (0, utils_1.getScrollableAncestors)(this.iFrame);
            this.lazyLoadScrollContainers.forEach((scrollContainer) => {
                scrollContainer.addEventListener('scroll', this.sendFullHeightLazyLoadData);
            });
            if (typeof ResizeObserver !== 'undefined') {
                const resizeTargets = new Set([
                    this.iFrame.parentElement,
                    ...(0, utils_1.getEffectiveClippingAncestors)(this.iFrame),
                ].filter(Boolean));
                this.lazyLoadResizeObserver = new ResizeObserver(this.sendFullHeightLazyLoadData);
                resizeTargets.forEach((resizeTarget) => {
                    this.lazyLoadResizeObserver.observe(resizeTarget);
                });
            }
        }
    }
    unregisterLazyLoadEvents() {
        var _a;
        if (this.viewConfig.fullHeight && this.viewConfig.lazyLoadingForFullHeight) {
            window.removeEventListener('resize', this.sendFullHeightLazyLoadData);
            window.removeEventListener('scroll', this.sendFullHeightLazyLoadData, true);
            (_a = this.lazyLoadResizeObserver) === null || _a === void 0 ? void 0 : _a.disconnect();
            this.lazyLoadResizeObserver = undefined;
            this.lazyLoadScrollContainers.forEach((scrollContainer) => {
                scrollContainer.removeEventListener('scroll', this.sendFullHeightLazyLoadData);
            });
            this.lazyLoadScrollContainers = [];
        }
    }
    /**
     * Renders the embedded application pages in the ThoughtSpot app.
     * @param renderOptions An object containing the page ID
     * to be embedded.
     */
    async render() {
        await super.render();
        const src = this.getIFrameSrc();
        await this.renderV1Embed(src);
        this.postRender();
        return this;
    }
}
exports.AppEmbed = AppEmbed;
