// @thoughtspot/visual-embed-sdk/cjs/src/embed/events.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const logger_1 = require("../utils/logger");
const index_1 = require("../index");
const test_utils_1 = require("../test/test-utils");
const authInstance = tslib_1.__importStar(require("../auth"));
const thoughtSpotHost = 'tshost';
const defaultViewConfig = {
    frameParams: {
        width: 1280,
        height: 720,
    },
};
const PAYLOAD = 'Sample payload';
beforeAll(() => {
    (0, index_1.init)({
        thoughtSpotHost,
        authType: index_1.AuthType.None,
    });
    jest.spyOn(window, 'alert');
    jest.spyOn(authInstance, 'postLoginService').mockImplementation(() => Promise.resolve(undefined));
});
describe('test communication between host app and ThoughtSpot', () => {
    beforeEach(() => {
        document.body.innerHTML = (0, test_utils_1.getDocumentBody)();
    });
    test('should capture event from ThoughtSpot app', (done) => {
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed
            .on(index_1.EmbedEvent.CustomAction, (data) => {
            expect(data.data).toBe(PAYLOAD);
            done();
        })
            .render();
        (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.CustomAction,
                data: PAYLOAD,
            });
        });
    });
    test('should capture SpotterResponseComplete event from ThoughtSpot app', (done) => {
        const appEmbed = new index_1.AppEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        appEmbed
            .on(index_1.EmbedEvent.SpotterResponseComplete, (data) => {
            expect(data.data).toEqual({ convId: 'c1', messageId: 'm1' });
            done();
        })
            .render();
        (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.SpotterResponseComplete,
                data: { convId: 'c1', messageId: 'm1' },
            });
        });
    });
    // TODO: enable test once we are actually able to load stuff in the iframe
    xtest('should trigger iframe load event', async () => {
        const onLoadSpy = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), {});
        searchEmbed.on(index_1.EmbedEvent.Load, onLoadSpy).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(onLoadSpy).toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('should trigger event to ThoughtSpot app', (done) => {
        (0, test_utils_1.mockMessageChannel)();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), {});
        searchEmbed.render();
        setTimeout(() => {
            searchEmbed.trigger(index_1.HostEvent.Search, {
                body: PAYLOAD,
            });
        }, test_utils_1.EVENT_WAIT_TIME);
        (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            iframe.contentWindow.addEventListener('message', (e) => {
                expect(e.data.type).toBe(index_1.HostEvent.Search);
                expect(e.data.data.body).toBe(PAYLOAD);
                done();
            });
        });
    });
    test('should execute multiple event handlers if registered', async () => {
        const handlerOne = jest.fn();
        const handlerTwo = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed
            .on(index_1.EmbedEvent.CustomAction, handlerOne)
            .on(index_1.EmbedEvent.CustomAction, handlerTwo)
            .render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.CustomAction,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handlerOne).toHaveBeenCalled();
            expect(handlerTwo).toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('should capture event from correct iframe', async () => {
        const spyOne = jest.fn();
        const embedOne = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        embedOne.on(index_1.EmbedEvent.CustomAction, spyOne).render();
        const spyTwo = jest.fn();
        const embedTwo = new index_1.PinboardEmbed((0, test_utils_1.getRootEl2)(), {
            ...defaultViewConfig,
            pinboardId: 'eca215d4-0d2c-4a55-90e3-d81ef6848ae0',
        });
        const spyThree = jest.fn();
        const embedThree = new index_1.LiveboardEmbed((0, test_utils_1.getRootEl2)(), {
            ...defaultViewConfig,
            liveboardId: 'eca215d4-0d2c-4a55-90e3-d81ef6848ae0',
        });
        embedTwo.on(index_1.EmbedEvent.CustomAction, spyTwo).render();
        embedThree.on(index_1.EmbedEvent.CustomAction, spyThree).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframeOne = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframeOne.contentWindow, {
                type: index_1.EmbedEvent.CustomAction,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(spyOne).toHaveBeenCalled();
            expect(spyTwo).not.toHaveBeenCalled();
            expect(spyThree).not.toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('send getIframeCenter Event without eventPort', async () => {
        const liveboardEmbed = new index_1.LiveboardEmbed((0, test_utils_1.getRootEl)(), {
            ...defaultViewConfig,
            fullHeight: true,
            pinboardId: 'eca215d4-0d2c-4a55-90e3-d81ef6848ae0',
        });
        liveboardEmbed.render();
        const spy1 = jest.spyOn(logger_1.logger, 'log');
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.EmbedIframeCenter,
                data: PAYLOAD,
            });
        });
        expect(spy1).toHaveBeenCalledWith('Event Port is not defined');
    });
    test('send getIframeCenter Event without eventPort - pinboard', async () => {
        const pinboardEmbed = new index_1.PinboardEmbed((0, test_utils_1.getRootEl)(), {
            ...defaultViewConfig,
            fullHeight: true,
            pinboardId: 'eca215d4-0d2c-4a55-90e3-d81ef6848ae0',
        });
        pinboardEmbed.render();
        const spy1 = jest.spyOn(logger_1.logger, 'log');
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.EmbedIframeCenter,
                data: PAYLOAD,
            });
        });
        expect(spy1).toHaveBeenCalledWith('Event Port is not defined');
    });
    test('send getIframeCenter Event with eventPort - pinboard', async () => {
        const pinboardEmbed = new index_1.PinboardEmbed((0, test_utils_1.getRootEl)(), {
            ...defaultViewConfig,
            fullHeight: true,
            pinboardId: 'eca215d4-0d2c-4a55-90e3-d81ef6848ae0',
        });
        pinboardEmbed.render();
        const mockPort = {
            postMessage: jest.fn(),
        };
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.EmbedIframeCenter,
                data: PAYLOAD,
            }, mockPort);
        });
        const heightObj = {
            data: {
                iframeCenter: 0,
                iframeHeight: 0,
                iframeScrolled: 0,
                iframeVisibleViewPort: 0,
                viewPortHeight: 768,
            },
            type: index_1.EmbedEvent.EmbedIframeCenter,
        };
        expect(mockPort.postMessage).toHaveBeenCalledWith(heightObj);
    });
    test('send getIframeCenter Event with eventPort', async () => {
        const liveboardEmbed = new index_1.LiveboardEmbed((0, test_utils_1.getRootEl)(), {
            ...defaultViewConfig,
            fullHeight: true,
            pinboardId: 'eca215d4-0d2c-4a55-90e3-d81ef6848ae0',
        });
        liveboardEmbed.render();
        const mockPort = {
            postMessage: jest.fn(),
        };
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.EmbedIframeCenter,
                data: PAYLOAD,
            }, mockPort);
        });
        const heightObj = {
            data: {
                iframeCenter: 0,
                iframeHeight: 0,
                iframeScrolled: 0,
                iframeVisibleViewPort: 0,
                viewPortHeight: 768,
            },
            type: index_1.EmbedEvent.EmbedIframeCenter,
        };
        expect(mockPort.postMessage).toHaveBeenCalledWith(heightObj);
    });
    test('should handle UpdatePersonalizedView as alias for UpdatePersonalisedView', async () => {
        const handler = jest.fn();
        const liveboardEmbed = new index_1.LiveboardEmbed((0, test_utils_1.getRootEl)(), {
            ...defaultViewConfig,
            liveboardId: 'eca215d4-0d2c-4a55-90e3-d81ef6848ae0',
        });
        liveboardEmbed.on(index_1.EmbedEvent.UpdatePersonalizedView, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.UpdatePersonalisedView,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data).toBe(PAYLOAD);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('should handle SavePersonalizedView as alias for SavePersonalisedView', async () => {
        const handler = jest.fn();
        const liveboardEmbed = new index_1.LiveboardEmbed((0, test_utils_1.getRootEl)(), {
            ...defaultViewConfig,
            liveboardId: 'eca215d4-0d2c-4a55-90e3-d81ef6848ae0',
        });
        liveboardEmbed.on(index_1.EmbedEvent.SavePersonalizedView, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.SavePersonalisedView,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data).toBe(PAYLOAD);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('should handle DeletePersonalizedView as alias for DeletePersonalisedView', async () => {
        const handler = jest.fn();
        const liveboardEmbed = new index_1.LiveboardEmbed((0, test_utils_1.getRootEl)(), {
            ...defaultViewConfig,
            liveboardId: 'eca215d4-0d2c-4a55-90e3-d81ef6848ae0',
        });
        liveboardEmbed.on(index_1.EmbedEvent.DeletePersonalizedView, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.DeletePersonalisedView,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data).toBe(PAYLOAD);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('should fire QueryChanged handler when {type:queryChanged} postMessage arrives from iframe', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.QueryChanged, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.QueryChanged,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data).toBe(PAYLOAD);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('should not fire QueryChanged handler for postMessage from non-iframe source', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.QueryChanged, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            (0, test_utils_1.postMessageToParent)(window, {
                type: index_1.EmbedEvent.QueryChanged,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).not.toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('ALL event listener should fire for all events with the event type set correctly', async () => {
        const embed = new index_1.AppEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        const spy = jest.fn();
        embed.on(index_1.EmbedEvent.ALL, spy);
        embed.render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.CustomAction,
                data: PAYLOAD,
            });
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.DialogOpen,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(spy).toHaveBeenCalledTimes(3);
            expect(spy.mock.calls[0][0]).toMatchObject({
                type: index_1.EmbedEvent.Init,
            });
            expect(spy.mock.calls[1][0]).toMatchObject({
                type: index_1.EmbedEvent.CustomAction,
                data: PAYLOAD,
            });
            expect(spy.mock.calls[2][0]).toMatchObject({
                type: index_1.EmbedEvent.DialogOpen,
            });
        }, test_utils_1.EVENT_WAIT_TIME);
    });
});
describe('Americanized enum alias values', () => {
    test('EmbedEvent aliases should resolve to the same values as their British counterparts', () => {
        expect(index_1.EmbedEvent.UpdatePersonalizedView).toBe(index_1.EmbedEvent.UpdatePersonalisedView);
        expect(index_1.EmbedEvent.SavePersonalizedView).toBe(index_1.EmbedEvent.SavePersonalisedView);
        expect(index_1.EmbedEvent.DeletePersonalizedView).toBe(index_1.EmbedEvent.DeletePersonalisedView);
    });
    test('HostEvent aliases should resolve to the same values as their British counterparts', () => {
        expect(index_1.HostEvent.ResetLiveboardPersonalizedView).toBe(index_1.HostEvent.ResetLiveboardPersonalisedView);
        expect(index_1.HostEvent.UpdatePersonalizedView).toBe(index_1.HostEvent.UpdatePersonalisedView);
    });
    test('Action aliases should resolve to the same values as their British counterparts', () => {
        expect(index_1.Action.PersonalizedViewsDropdown).toBe(index_1.Action.PersonalisedViewsDropdown);
        expect(index_1.Action.OrganizeFavorites).toBe(index_1.Action.OrganiseFavourites);
    });
});
describe('Action enum values', () => {
    test('RemoveFromFavorites should resolve to its string value', () => {
        expect(index_1.Action.RemoveFromFavorites).toBe('removeFromFavorites');
    });
});
describe('EmbedEvent basic routing', () => {
    beforeEach(() => {
        document.body.innerHTML = (0, test_utils_1.getDocumentBody)();
    });
    const UNCOVERED_EVENTS = [
        ['Drilldown', index_1.EmbedEvent.Drilldown],
        ['DataSourceSelected', index_1.EmbedEvent.DataSourceSelected],
        ['AddRemoveColumns', index_1.EmbedEvent.AddRemoveColumns],
        ['VizPointDoubleClick', index_1.EmbedEvent.VizPointDoubleClick],
        ['VizPointClick', index_1.EmbedEvent.VizPointClick],
        ['Alert', index_1.EmbedEvent.Alert],
        ['GetDataClick', index_1.EmbedEvent.GetDataClick],
        ['DialogClose', index_1.EmbedEvent.DialogClose],
        ['Download', index_1.EmbedEvent.Download],
        ['DownloadAsPng', index_1.EmbedEvent.DownloadAsPng],
        ['DownloadAsPdf', index_1.EmbedEvent.DownloadAsPdf],
        ['DownloadAsCsv', index_1.EmbedEvent.DownloadAsCsv],
        ['DownloadAsXlsx', index_1.EmbedEvent.DownloadAsXlsx],
        ['DownloadLiveboardAsContinuousPDF', index_1.EmbedEvent.DownloadLiveboardAsContinuousPDF],
        ['AnswerDelete', index_1.EmbedEvent.AnswerDelete],
        ['AIHighlights', index_1.EmbedEvent.AIHighlights],
        ['Pin', index_1.EmbedEvent.Pin],
        ['SpotIQAnalyze', index_1.EmbedEvent.SpotIQAnalyze],
        ['Share', index_1.EmbedEvent.Share],
        ['DrillInclude', index_1.EmbedEvent.DrillInclude],
        ['DrillExclude', index_1.EmbedEvent.DrillExclude],
        ['CopyToClipboard', index_1.EmbedEvent.CopyToClipboard],
        ['UpdateTML', index_1.EmbedEvent.UpdateTML],
        ['EditTML', index_1.EmbedEvent.EditTML],
        ['ExportTML', index_1.EmbedEvent.ExportTML],
        ['SaveAsView', index_1.EmbedEvent.SaveAsView],
        ['CopyAEdit', index_1.EmbedEvent.CopyAEdit],
        ['ShowUnderlyingData', index_1.EmbedEvent.ShowUnderlyingData],
        ['AnswerChartSwitcher', index_1.EmbedEvent.AnswerChartSwitcher],
        ['LiveboardInfo', index_1.EmbedEvent.LiveboardInfo],
        ['AddToFavorites', index_1.EmbedEvent.AddToFavorites],
        ['Schedule', index_1.EmbedEvent.Schedule],
        ['Edit', index_1.EmbedEvent.Edit],
        ['MakeACopy', index_1.EmbedEvent.MakeACopy],
        ['Present', index_1.EmbedEvent.Present],
        ['Delete', index_1.EmbedEvent.Delete],
        ['SchedulesList', index_1.EmbedEvent.SchedulesList],
        ['Cancel', index_1.EmbedEvent.Cancel],
        ['Explore', index_1.EmbedEvent.Explore],
        ['CopyLink', index_1.EmbedEvent.CopyLink],
        ['CrossFilterChanged', index_1.EmbedEvent.CrossFilterChanged],
        ['VizPointRightClick', index_1.EmbedEvent.VizPointRightClick],
        ['InsertIntoSlide', index_1.EmbedEvent.InsertIntoSlide],
        ['FilterChanged', index_1.EmbedEvent.FilterChanged],
        ['UpdateConnection', index_1.EmbedEvent.UpdateConnection],
        ['CreateConnection', index_1.EmbedEvent.CreateConnection],
        ['ResetLiveboard', index_1.EmbedEvent.ResetLiveboard],
        ['ChangePersonalizedView', index_1.EmbedEvent.ChangePersonalizedView],
        ['CreateWorksheet', index_1.EmbedEvent.CreateWorksheet],
        ['AskSageInit', index_1.EmbedEvent.AskSageInit],
        ['Rename', index_1.EmbedEvent.Rename],
        ['ParameterChanged', index_1.EmbedEvent.ParameterChanged],
        ['TableVizRendered', index_1.EmbedEvent.TableVizRendered],
        ['CreateLiveboard', index_1.EmbedEvent.CreateLiveboard],
        ['CreateModel', index_1.EmbedEvent.CreateModel],
        ['SpotterData', index_1.EmbedEvent.SpotterData],
        ['PreviewSpotterData', index_1.EmbedEvent.PreviewSpotterData],
        ['AddToCoaching', index_1.EmbedEvent.AddToCoaching],
        ['DataModelInstructions', index_1.EmbedEvent.DataModelInstructions],
        ['SpotterQueryTriggered', index_1.EmbedEvent.SpotterQueryTriggered],
        ['LastPromptEdited', index_1.EmbedEvent.LastPromptEdited],
        ['LastPromptDeleted', index_1.EmbedEvent.LastPromptDeleted],
        ['ResetSpotterConversation', index_1.EmbedEvent.ResetSpotterConversation],
        ['SpotterInit', index_1.EmbedEvent.SpotterInit],
        ['SpotterLoadComplete', index_1.EmbedEvent.SpotterLoadComplete],
        ['OrgSwitched', index_1.EmbedEvent.OrgSwitched],
        ['SpotterConversationRenamed', index_1.EmbedEvent.SpotterConversationRenamed],
        ['SpotterConversationDeleted', index_1.EmbedEvent.SpotterConversationDeleted],
        ['SpotterConversationSelected', index_1.EmbedEvent.SpotterConversationSelected],
        ['EmbedPageContextChanged', index_1.EmbedEvent.EmbedPageContextChanged],
        ['Subscribed', index_1.EmbedEvent.Subscribed],
        ['SendTestScheduleEmail', index_1.EmbedEvent.SendTestScheduleEmail],
        ['SpotterVizInit', index_1.EmbedEvent.SpotterVizInit],
        ['SpotterVizQueryTriggered', index_1.EmbedEvent.SpotterVizQueryTriggered],
        ['SpotterVizResponseComplete', index_1.EmbedEvent.SpotterVizResponseComplete],
        ['SpotterVizCheckpointCreated', index_1.EmbedEvent.SpotterVizCheckpointCreated],
        ['SpotterVizCheckpointRestored', index_1.EmbedEvent.SpotterVizCheckpointRestored],
        ['SpotterVizError', index_1.EmbedEvent.SpotterVizError],
        ['SpotterVizClosed', index_1.EmbedEvent.SpotterVizClosed],
        ['RefreshLiveboardBrowserCache', index_1.EmbedEvent.RefreshLiveboardBrowserCache],
        ['V1Data', index_1.EmbedEvent.V1Data],
    ];
    test.each(UNCOVERED_EVENTS)('%s handler fires on iframe postMessage', async (eventName, embedEvent) => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(embedEvent, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: embedEvent,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data).toBe(PAYLOAD);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
});
describe('EmbedEvent start/end status', () => {
    beforeEach(() => {
        document.body.innerHTML = (0, test_utils_1.getDocumentBody)();
    });
    const START_END_EVENTS = [
        ['DownloadAsPng', index_1.EmbedEvent.DownloadAsPng],
        ['DownloadAsPdf', index_1.EmbedEvent.DownloadAsPdf],
        ['DownloadAsCsv', index_1.EmbedEvent.DownloadAsCsv],
        ['DownloadAsXlsx', index_1.EmbedEvent.DownloadAsXlsx],
        ['AnswerDelete', index_1.EmbedEvent.AnswerDelete],
        ['Pin', index_1.EmbedEvent.Pin],
        ['SpotIQAnalyze', index_1.EmbedEvent.SpotIQAnalyze],
        ['Share', index_1.EmbedEvent.Share],
        ['ExportTML', index_1.EmbedEvent.ExportTML],
        ['CopyAEdit', index_1.EmbedEvent.CopyAEdit],
    ];
    test.each(START_END_EVENTS)('%s start handler fires on START, not END', async (eventName, embedEvent) => {
        const startHandler = jest.fn();
        const endHandler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed
            .on(embedEvent, startHandler, { start: true })
            .on(embedEvent, endHandler)
            .render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: embedEvent,
                data: PAYLOAD,
                status: 'start',
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(startHandler).toHaveBeenCalled();
            expect(endHandler).not.toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test.each(START_END_EVENTS)('%s end handler fires on END, not START', async (eventName, embedEvent) => {
        const startHandler = jest.fn();
        const endHandler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed
            .on(embedEvent, startHandler, { start: true })
            .on(embedEvent, endHandler)
            .render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: embedEvent,
                data: PAYLOAD,
                status: 'end',
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(endHandler).toHaveBeenCalled();
            expect(startHandler).not.toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    const END_ONLY_EVENTS = [
        ['Download', index_1.EmbedEvent.Download],
        ['DownloadLiveboardAsContinuousPDF', index_1.EmbedEvent.DownloadLiveboardAsContinuousPDF],
    ];
    test.each(END_ONLY_EVENTS)('%s ignores start handler (end-only event)', async (eventName, embedEvent) => {
        const startHandler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(embedEvent, startHandler, { start: true }).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: embedEvent,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(startHandler).not.toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
});
describe('EmbedEvent payload and routing', () => {
    beforeEach(() => {
        document.body.innerHTML = (0, test_utils_1.getDocumentBody)();
    });
    test('Drilldown payload includes additionalFilters, drillDownColumns, nonFilteredColumns', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.Drilldown, handler).render();
        const drilldownData = {
            additionalFilters: [{ column: 'region', value: 'west' }],
            drillDownColumns: ['revenue'],
            nonFilteredColumns: ['date'],
        };
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.Drilldown,
                data: drilldownData,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            const received = handler.mock.calls[0][0].data;
            expect(received.additionalFilters).toEqual(drilldownData.additionalFilters);
            expect(received.drillDownColumns).toEqual(drilldownData.drillDownColumns);
            expect(received.nonFilteredColumns).toEqual(drilldownData.nonFilteredColumns);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('DataSourceSelected payload contains dataSourceIds array', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.DataSourceSelected, handler).render();
        const dataSourceIds = ['ds-aaa', 'ds-bbb'];
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.DataSourceSelected,
                data: { dataSourceIds },
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data.dataSourceIds).toEqual(dataSourceIds);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('V1Data fires via __type field when type is absent', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.V1Data, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                __type: index_1.EmbedEvent.V1Data,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('subscribedEvent builds composed key and routes to registered handler', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        const composedEvent = searchEmbed.subscribedEvent(index_1.HostEvent.Save);
        expect(composedEvent).toBe(`${index_1.HostEvent.Save} ${index_1.EmbedEvent.Subscribed}`);
        searchEmbed.on(composedEvent, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: composedEvent,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('non-iframe postMessage does not trigger registered handler', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.Drilldown, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            (0, test_utils_1.postMessageToParent)(window, {
                type: index_1.EmbedEvent.Drilldown,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).not.toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('DataSourceSelected falls back to payload field when data is absent', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.DataSourceSelected, handler).render();
        const dataSourceIds = ['ds-fallback'];
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.DataSourceSelected,
                payload: { dataSourceIds },
                // no data field — formatEventData should copy payload → data
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data.dataSourceIds).toEqual(dataSourceIds);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('Alert handler fires independently of window.alert', async () => {
        const alertSpy = jest.spyOn(window, 'alert').mockImplementation(() => undefined);
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.Alert, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.Alert,
                data: { message: 'test alert' },
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(alertSpy).not.toHaveBeenCalled();
        }, test_utils_1.EVENT_WAIT_TIME);
        alertSpy.mockRestore();
    });
    test('Share reaches EmbedEvent.ALL wildcard handler', async () => {
        const allHandler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.ALL, allHandler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.Share,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            const types = allHandler.mock.calls.map((c) => c[0].type);
            expect(types).toContain(index_1.EmbedEvent.Share);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('DialogClose reaches EmbedEvent.ALL wildcard handler', async () => {
        const allHandler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.ALL, allHandler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.DialogClose,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            const types = allHandler.mock.calls.map((c) => c[0].type);
            expect(types).toContain(index_1.EmbedEvent.DialogClose);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('SpotterConversationRenamed payload contains convId, oldTitle, newTitle', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.SpotterConversationRenamed, handler).render();
        const convData = { convId: 'conv-1', oldTitle: 'Old', newTitle: 'New' };
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.SpotterConversationRenamed,
                data: convData,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data).toMatchObject(convData);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('SpotterConversationDeleted payload contains convId and title', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.SpotterConversationDeleted, handler).render();
        const convData = { convId: 'conv-2', title: 'My Conversation' };
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.SpotterConversationDeleted,
                data: convData,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data).toMatchObject(convData);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('SpotterConversationSelected payload contains convId, title, worksheetId', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.SpotterConversationSelected, handler).render();
        const convData = { convId: 'conv-3', title: 'My Conv', worksheetId: 'ws-1' };
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.SpotterConversationSelected,
                data: convData,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data).toMatchObject(convData);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('TableVizRendered payload forwards columnDataLite array intact', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.TableVizRendered, handler).render();
        const columnDataLite = [['col1', 'col2'], ['val1', 'val2']];
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.TableVizRendered,
                data: { data: { columnDataLite } },
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data.data.columnDataLite).toEqual(columnDataLite);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('SendTestScheduleEmail payload preserves liveboardId and sendToSelf boolean', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.SendTestScheduleEmail, handler).render();
        const emailData = { liveboardId: 'lb-123', sendToSelf: true };
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.SendTestScheduleEmail,
                data: emailData,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            const received = handler.mock.calls[0][0].data;
            expect(received.liveboardId).toBe('lb-123');
            expect(received.sendToSelf).toBe(true);
            expect(typeof received.sendToSelf).toBe('boolean');
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('ChangePersonalizedView payload viewId can be null (reset-to-default)', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.ChangePersonalizedView, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.ChangePersonalizedView,
                data: { viewId: null, viewName: 'Original View', liveboardId: 'lb-1', isPublic: false },
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            expect(handler.mock.calls[0][0].data.viewId).toBeNull();
            expect(handler.mock.calls[0][0].data.viewName).toBe('Original View');
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('VizPointDoubleClick payload exposes vizId and clickedPoint.selectedPoints', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.VizPointDoubleClick, handler).render();
        const pointData = {
            vizId: 'viz-1',
            clickedPoint: { selectedPoints: [{ x: 10, y: 20 }] },
        };
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.VizPointDoubleClick,
                data: pointData,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            const { data } = handler.mock.calls[0][0];
            expect(data.vizId).toBe('viz-1');
            expect(data.clickedPoint.selectedPoints).toEqual(pointData.clickedPoint.selectedPoints);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('VizPointClick payload exposes vizId, selectedAttributes and selectedMeasures', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.VizPointClick, handler).render();
        const pointData = {
            vizId: 'viz-2',
            clickedPoint: {
                selectedAttributes: [{ value: 'West', column: { name: 'region' } }],
                selectedMeasures: [{ value: 42, column: { name: 'revenue' } }],
            },
        };
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.VizPointClick,
                data: pointData,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            const { data } = handler.mock.calls[0][0];
            expect(data.vizId).toBe('viz-2');
            expect(data.clickedPoint.selectedAttributes[0].value).toBe('West');
            expect(data.clickedPoint.selectedAttributes[0].column.name).toBe('region');
            expect(data.clickedPoint.selectedMeasures[0].value).toBe(42);
            expect(data.clickedPoint.selectedMeasures[0].column.name).toBe('revenue');
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('SpotterVizError reaches EmbedEvent.ALL handler', async () => {
        const allHandler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.ALL, allHandler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.SpotterVizError,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            const types = allHandler.mock.calls.map((c) => c[0].type);
            expect(types).toContain(index_1.EmbedEvent.SpotterVizError);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('SpotterVizClosed reaches EmbedEvent.ALL handler', async () => {
        const allHandler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.ALL, allHandler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.SpotterVizClosed,
                data: PAYLOAD,
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            const types = allHandler.mock.calls.map((c) => c[0].type);
            expect(types).toContain(index_1.EmbedEvent.SpotterVizClosed);
        }, test_utils_1.EVENT_WAIT_TIME);
    });
    test('SpotterVizCheckpointRestored payload newGenNumber is a number', async () => {
        const handler = jest.fn();
        const searchEmbed = new index_1.SearchEmbed((0, test_utils_1.getRootEl)(), defaultViewConfig);
        searchEmbed.on(index_1.EmbedEvent.SpotterVizCheckpointRestored, handler).render();
        await (0, test_utils_1.executeAfterWait)(() => {
            const iframe = (0, test_utils_1.getIFrameEl)();
            (0, test_utils_1.postMessageToParent)(iframe.contentWindow, {
                type: index_1.EmbedEvent.SpotterVizCheckpointRestored,
                data: { checkpointId: 'cp-1', newGenNumber: 42 },
            });
        });
        await (0, test_utils_1.executeAfterWait)(() => {
            expect(handler).toHaveBeenCalled();
            const { newGenNumber } = handler.mock.calls[0][0].data;
            expect(newGenNumber).toBe(42);
            expect(typeof newGenNumber).toBe('number');
        }, test_utils_1.EVENT_WAIT_TIME);
    });
});
