// @thoughtspot/visual-embed-sdk/cjs/src/embed/spotter-viz-utils.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const spotter_viz_utils_1 = require("./spotter-viz-utils");
describe('buildSpotterVizAppInitData', () => {
    const base = { type: 'APP_INIT' };
    it('returns initData unchanged when spotterViz is not provided', () => {
        const result = (0, spotter_viz_utils_1.buildSpotterVizAppInitData)(base, {});
        expect(result).toBe(base);
    });
    it('nests spotterViz under embedParams.spotterVizConfig', () => {
        var _a;
        const spotterViz = { brandName: 'MyBrand', description: 'Desc', inputChatPlaceholder: 'Ask...' };
        const result = (0, spotter_viz_utils_1.buildSpotterVizAppInitData)(base, { spotterViz });
        expect((_a = result.embedParams) === null || _a === void 0 ? void 0 : _a.spotterVizConfig).toEqual(spotterViz);
    });
    it('passes brandHeadline through spotterVizConfig', () => {
        var _a, _b;
        const spotterViz = { brandName: 'MyBrand', brandHeadline: "Hi, there! I'm" };
        const result = (0, spotter_viz_utils_1.buildSpotterVizAppInitData)(base, { spotterViz });
        expect((_b = (_a = result.embedParams) === null || _a === void 0 ? void 0 : _a.spotterVizConfig) === null || _b === void 0 ? void 0 : _b.brandHeadline).toBe("Hi, there! I'm");
    });
    it('passes liveboardBrandName, spotterBrandName, insightTileBrandName, insightTileViewPlanLabel and insightTileLoaderText through spotterVizConfig', () => {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        const spotterViz = {
            brandName: 'MyBrand',
            liveboardBrandName: 'Reports',
            spotterBrandName: 'Analyst',
            insightTileBrandName: 'Insight card',
            insightTileViewPlanLabel: 'View plan',
            insightTileLoaderText: 'Generating insight',
        };
        const result = (0, spotter_viz_utils_1.buildSpotterVizAppInitData)(base, { spotterViz });
        expect((_b = (_a = result.embedParams) === null || _a === void 0 ? void 0 : _a.spotterVizConfig) === null || _b === void 0 ? void 0 : _b.liveboardBrandName).toBe('Reports');
        expect((_d = (_c = result.embedParams) === null || _c === void 0 ? void 0 : _c.spotterVizConfig) === null || _d === void 0 ? void 0 : _d.spotterBrandName).toBe('Analyst');
        expect((_f = (_e = result.embedParams) === null || _e === void 0 ? void 0 : _e.spotterVizConfig) === null || _f === void 0 ? void 0 : _f.insightTileBrandName).toBe('Insight card');
        expect((_h = (_g = result.embedParams) === null || _g === void 0 ? void 0 : _g.spotterVizConfig) === null || _h === void 0 ? void 0 : _h.insightTileViewPlanLabel).toBe('View plan');
        expect((_k = (_j = result.embedParams) === null || _j === void 0 ? void 0 : _j.spotterVizConfig) === null || _k === void 0 ? void 0 : _k.insightTileLoaderText).toBe('Generating insight');
    });
    it('preserves existing embedParams when adding spotterVizConfig', () => {
        var _a, _b, _c;
        const existing = { ...base, embedParams: { spotterSidebarConfig: { enablePastConversationsSidebar: true } } };
        const spotterViz = { brandName: 'MyBrand' };
        const result = (0, spotter_viz_utils_1.buildSpotterVizAppInitData)(existing, { spotterViz });
        expect((_a = result.embedParams) === null || _a === void 0 ? void 0 : _a.spotterVizConfig).toEqual(spotterViz);
        expect((_c = (_b = result.embedParams) === null || _b === void 0 ? void 0 : _b.spotterSidebarConfig) === null || _c === void 0 ? void 0 : _c.enablePastConversationsSidebar).toBe(true);
    });
});
