// @thoughtspot/visual-embed-sdk/cjs/src/react/util.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
const types_1 = require("../types");
describe('React util functions', () => {
    describe('getViewPropsAndListeners', () => {
        test('should return empty viewConfig and listeners for empty props', () => {
            const props = {};
            const result = (0, util_1.getViewPropsAndListeners)(props);
            expect(result.viewConfig).toEqual({});
            expect(result.listeners).toEqual({});
        });
        test('should separate view config properties from props', () => {
            const props = {
                frameParams: { width: 100, height: 200 },
                showLiveboardTitle: true,
                liveboardId: 'test-liveboard-id',
                vizId: 'test-viz-id',
                className: 'test-class',
                style: { color: 'red' },
            };
            const result = (0, util_1.getViewPropsAndListeners)(props);
            expect(result.viewConfig).toEqual({
                frameParams: { width: 100, height: 200 },
                showLiveboardTitle: true,
                liveboardId: 'test-liveboard-id',
                vizId: 'test-viz-id',
                className: 'test-class',
                style: { color: 'red' },
            });
            expect(result.listeners).toEqual({});
        });
        test('should separate event handlers from props', () => {
            const onInit = jest.fn();
            const onLoad = jest.fn();
            const onData = jest.fn();
            const props = {
                onInit,
                onLoad,
                onData,
            };
            const result = (0, util_1.getViewPropsAndListeners)(props);
            expect(result.viewConfig).toEqual({});
            expect(result.listeners).toEqual({
                [types_1.EmbedEvent.Init]: onInit,
                [types_1.EmbedEvent.Load]: onLoad,
                [types_1.EmbedEvent.Data]: onData,
            });
        });
        test('should handle both view config and event handlers', () => {
            const onInit = jest.fn();
            const onAuthInit = jest.fn();
            const onQueryChanged = jest.fn();
            const props = {
                liveboardId: 'test-liveboard-id',
                showLiveboardTitle: false,
                frameParams: { height: 500 },
                onInit,
                onAuthInit,
                onQueryChanged,
                className: 'embed-container',
            };
            const result = (0, util_1.getViewPropsAndListeners)(props);
            expect(result.viewConfig).toEqual({
                liveboardId: 'test-liveboard-id',
                showLiveboardTitle: false,
                frameParams: { height: 500 },
                className: 'embed-container',
            });
            expect(result.listeners).toEqual({
                [types_1.EmbedEvent.Init]: onInit,
                [types_1.EmbedEvent.AuthInit]: onAuthInit,
                [types_1.EmbedEvent.QueryChanged]: onQueryChanged,
            });
        });
        test('all onX props map to their correct EmbedEvent listener keys', () => {
            const onDrilldown = jest.fn();
            const onDataSourceSelected = jest.fn();
            const onAddRemoveColumns = jest.fn();
            const onVizPointDoubleClick = jest.fn();
            const onVizPointClick = jest.fn();
            const onAlert = jest.fn();
            const onGetDataClick = jest.fn();
            const onDialogClose = jest.fn();
            const onDownload = jest.fn();
            const onDownloadAsPng = jest.fn();
            const onDownloadAsPdf = jest.fn();
            const onDownloadAsCsv = jest.fn();
            const onDownloadAsXlsx = jest.fn();
            const onDownloadLiveboardAsContinuousPDF = jest.fn();
            const onAnswerDelete = jest.fn();
            const onAIHighlights = jest.fn();
            const onPin = jest.fn();
            const onSpotIQAnalyze = jest.fn();
            const onShare = jest.fn();
            const onDrillInclude = jest.fn();
            const onDrillExclude = jest.fn();
            const onCopyToClipboard = jest.fn();
            const onUpdateTML = jest.fn();
            const onEditTML = jest.fn();
            const onExportTML = jest.fn();
            const onSaveAsView = jest.fn();
            const onCopyAEdit = jest.fn();
            const onShowUnderlyingData = jest.fn();
            const onAnswerChartSwitcher = jest.fn();
            const onLiveboardInfo = jest.fn();
            const onAddToFavorites = jest.fn();
            const onSchedule = jest.fn();
            const onEdit = jest.fn();
            const onMakeACopy = jest.fn();
            const onPresent = jest.fn();
            const onDelete = jest.fn();
            const onSchedulesList = jest.fn();
            const onCancel = jest.fn();
            const onExplore = jest.fn();
            const onCopyLink = jest.fn();
            const onCrossFilterChanged = jest.fn();
            const onVizPointRightClick = jest.fn();
            const onInsertIntoSlide = jest.fn();
            const onFilterChanged = jest.fn();
            const onUpdateConnection = jest.fn();
            const onCreateConnection = jest.fn();
            const onResetLiveboard = jest.fn();
            const onChangePersonalizedView = jest.fn();
            const onCreateWorksheet = jest.fn();
            const onAskSageInit = jest.fn();
            const onRename = jest.fn();
            const onParameterChanged = jest.fn();
            const onTableVizRendered = jest.fn();
            const onCreateLiveboard = jest.fn();
            const onCreateModel = jest.fn();
            const onSpotterData = jest.fn();
            const onPreviewSpotterData = jest.fn();
            const onAddToCoaching = jest.fn();
            const onDataModelInstructions = jest.fn();
            const onSpotterQueryTriggered = jest.fn();
            const onLastPromptEdited = jest.fn();
            const onLastPromptDeleted = jest.fn();
            const onResetSpotterConversation = jest.fn();
            const onSpotterInit = jest.fn();
            const onSpotterLoadComplete = jest.fn();
            const onOrgSwitched = jest.fn();
            const onSpotterConversationRenamed = jest.fn();
            const onSpotterConversationDeleted = jest.fn();
            const onSpotterConversationSelected = jest.fn();
            const onEmbedPageContextChanged = jest.fn();
            const onSubscribed = jest.fn();
            const onSendTestScheduleEmail = jest.fn();
            const onSpotterVizInit = jest.fn();
            const onSpotterVizQueryTriggered = jest.fn();
            const onSpotterVizResponseComplete = jest.fn();
            const onSpotterVizCheckpointCreated = jest.fn();
            const onSpotterVizCheckpointRestored = jest.fn();
            const onSpotterVizError = jest.fn();
            const onSpotterVizClosed = jest.fn();
            const onRefreshLiveboardBrowserCache = jest.fn();
            const onV1Data = jest.fn();
            const props = {
                onDrilldown,
                onDataSourceSelected,
                onAddRemoveColumns,
                onVizPointDoubleClick,
                onVizPointClick,
                onAlert,
                onGetDataClick,
                onDialogClose,
                onDownload,
                onDownloadAsPng,
                onDownloadAsPdf,
                onDownloadAsCsv,
                onDownloadAsXlsx,
                onDownloadLiveboardAsContinuousPDF,
                onAnswerDelete,
                onAIHighlights,
                onPin,
                onSpotIQAnalyze,
                onShare,
                onDrillInclude,
                onDrillExclude,
                onCopyToClipboard,
                onUpdateTML,
                onEditTML,
                onExportTML,
                onSaveAsView,
                onCopyAEdit,
                onShowUnderlyingData,
                onAnswerChartSwitcher,
                onLiveboardInfo,
                onAddToFavorites,
                onSchedule,
                onEdit,
                onMakeACopy,
                onPresent,
                onDelete,
                onSchedulesList,
                onCancel,
                onExplore,
                onCopyLink,
                onCrossFilterChanged,
                onVizPointRightClick,
                onInsertIntoSlide,
                onFilterChanged,
                onUpdateConnection,
                onCreateConnection,
                onResetLiveboard,
                onChangePersonalizedView,
                onCreateWorksheet,
                onAskSageInit,
                onRename,
                onParameterChanged,
                onTableVizRendered,
                onCreateLiveboard,
                onCreateModel,
                onSpotterData,
                onPreviewSpotterData,
                onAddToCoaching,
                onDataModelInstructions,
                onSpotterQueryTriggered,
                onLastPromptEdited,
                onLastPromptDeleted,
                onResetSpotterConversation,
                onSpotterInit,
                onSpotterLoadComplete,
                onOrgSwitched,
                onSpotterConversationRenamed,
                onSpotterConversationDeleted,
                onSpotterConversationSelected,
                onEmbedPageContextChanged,
                onSubscribed,
                onSendTestScheduleEmail,
                onSpotterVizInit,
                onSpotterVizQueryTriggered,
                onSpotterVizResponseComplete,
                onSpotterVizCheckpointCreated,
                onSpotterVizCheckpointRestored,
                onSpotterVizError,
                onSpotterVizClosed,
                onRefreshLiveboardBrowserCache,
                onV1Data,
            };
            const result = (0, util_1.getViewPropsAndListeners)(props);
            expect(result.viewConfig).toEqual({});
            expect(result.listeners).toEqual({
                [types_1.EmbedEvent.Drilldown]: onDrilldown,
                [types_1.EmbedEvent.DataSourceSelected]: onDataSourceSelected,
                [types_1.EmbedEvent.AddRemoveColumns]: onAddRemoveColumns,
                [types_1.EmbedEvent.VizPointDoubleClick]: onVizPointDoubleClick,
                [types_1.EmbedEvent.VizPointClick]: onVizPointClick,
                [types_1.EmbedEvent.Alert]: onAlert,
                [types_1.EmbedEvent.GetDataClick]: onGetDataClick,
                [types_1.EmbedEvent.DialogClose]: onDialogClose,
                [types_1.EmbedEvent.Download]: onDownload,
                [types_1.EmbedEvent.DownloadAsPng]: onDownloadAsPng,
                [types_1.EmbedEvent.DownloadAsPdf]: onDownloadAsPdf,
                [types_1.EmbedEvent.DownloadAsCsv]: onDownloadAsCsv,
                [types_1.EmbedEvent.DownloadAsXlsx]: onDownloadAsXlsx,
                [types_1.EmbedEvent.DownloadLiveboardAsContinuousPDF]: onDownloadLiveboardAsContinuousPDF,
                [types_1.EmbedEvent.AnswerDelete]: onAnswerDelete,
                [types_1.EmbedEvent.AIHighlights]: onAIHighlights,
                [types_1.EmbedEvent.Pin]: onPin,
                [types_1.EmbedEvent.SpotIQAnalyze]: onSpotIQAnalyze,
                [types_1.EmbedEvent.Share]: onShare,
                [types_1.EmbedEvent.DrillInclude]: onDrillInclude,
                [types_1.EmbedEvent.DrillExclude]: onDrillExclude,
                [types_1.EmbedEvent.CopyToClipboard]: onCopyToClipboard,
                [types_1.EmbedEvent.UpdateTML]: onUpdateTML,
                [types_1.EmbedEvent.EditTML]: onEditTML,
                [types_1.EmbedEvent.ExportTML]: onExportTML,
                [types_1.EmbedEvent.SaveAsView]: onSaveAsView,
                [types_1.EmbedEvent.CopyAEdit]: onCopyAEdit,
                [types_1.EmbedEvent.ShowUnderlyingData]: onShowUnderlyingData,
                [types_1.EmbedEvent.AnswerChartSwitcher]: onAnswerChartSwitcher,
                [types_1.EmbedEvent.LiveboardInfo]: onLiveboardInfo,
                [types_1.EmbedEvent.AddToFavorites]: onAddToFavorites,
                [types_1.EmbedEvent.Schedule]: onSchedule,
                [types_1.EmbedEvent.Edit]: onEdit,
                [types_1.EmbedEvent.MakeACopy]: onMakeACopy,
                [types_1.EmbedEvent.Present]: onPresent,
                [types_1.EmbedEvent.Delete]: onDelete,
                [types_1.EmbedEvent.SchedulesList]: onSchedulesList,
                [types_1.EmbedEvent.Cancel]: onCancel,
                [types_1.EmbedEvent.Explore]: onExplore,
                [types_1.EmbedEvent.CopyLink]: onCopyLink,
                [types_1.EmbedEvent.CrossFilterChanged]: onCrossFilterChanged,
                [types_1.EmbedEvent.VizPointRightClick]: onVizPointRightClick,
                [types_1.EmbedEvent.InsertIntoSlide]: onInsertIntoSlide,
                [types_1.EmbedEvent.FilterChanged]: onFilterChanged,
                [types_1.EmbedEvent.UpdateConnection]: onUpdateConnection,
                [types_1.EmbedEvent.CreateConnection]: onCreateConnection,
                [types_1.EmbedEvent.ResetLiveboard]: onResetLiveboard,
                [types_1.EmbedEvent.ChangePersonalizedView]: onChangePersonalizedView,
                [types_1.EmbedEvent.CreateWorksheet]: onCreateWorksheet,
                [types_1.EmbedEvent.AskSageInit]: onAskSageInit,
                [types_1.EmbedEvent.Rename]: onRename,
                [types_1.EmbedEvent.ParameterChanged]: onParameterChanged,
                [types_1.EmbedEvent.TableVizRendered]: onTableVizRendered,
                [types_1.EmbedEvent.CreateLiveboard]: onCreateLiveboard,
                [types_1.EmbedEvent.CreateModel]: onCreateModel,
                [types_1.EmbedEvent.SpotterData]: onSpotterData,
                [types_1.EmbedEvent.PreviewSpotterData]: onPreviewSpotterData,
                [types_1.EmbedEvent.AddToCoaching]: onAddToCoaching,
                [types_1.EmbedEvent.DataModelInstructions]: onDataModelInstructions,
                [types_1.EmbedEvent.SpotterQueryTriggered]: onSpotterQueryTriggered,
                [types_1.EmbedEvent.LastPromptEdited]: onLastPromptEdited,
                [types_1.EmbedEvent.LastPromptDeleted]: onLastPromptDeleted,
                [types_1.EmbedEvent.ResetSpotterConversation]: onResetSpotterConversation,
                [types_1.EmbedEvent.SpotterInit]: onSpotterInit,
                [types_1.EmbedEvent.SpotterLoadComplete]: onSpotterLoadComplete,
                [types_1.EmbedEvent.OrgSwitched]: onOrgSwitched,
                [types_1.EmbedEvent.SpotterConversationRenamed]: onSpotterConversationRenamed,
                [types_1.EmbedEvent.SpotterConversationDeleted]: onSpotterConversationDeleted,
                [types_1.EmbedEvent.SpotterConversationSelected]: onSpotterConversationSelected,
                [types_1.EmbedEvent.EmbedPageContextChanged]: onEmbedPageContextChanged,
                [types_1.EmbedEvent.Subscribed]: onSubscribed,
                [types_1.EmbedEvent.SendTestScheduleEmail]: onSendTestScheduleEmail,
                [types_1.EmbedEvent.SpotterVizInit]: onSpotterVizInit,
                [types_1.EmbedEvent.SpotterVizQueryTriggered]: onSpotterVizQueryTriggered,
                [types_1.EmbedEvent.SpotterVizResponseComplete]: onSpotterVizResponseComplete,
                [types_1.EmbedEvent.SpotterVizCheckpointCreated]: onSpotterVizCheckpointCreated,
                [types_1.EmbedEvent.SpotterVizCheckpointRestored]: onSpotterVizCheckpointRestored,
                [types_1.EmbedEvent.SpotterVizError]: onSpotterVizError,
                [types_1.EmbedEvent.SpotterVizClosed]: onSpotterVizClosed,
                [types_1.EmbedEvent.RefreshLiveboardBrowserCache]: onRefreshLiveboardBrowserCache,
                [types_1.EmbedEvent.V1Data]: onV1Data,
            });
        });
    });
});
