// @thoughtspot/visual-embed-sdk/cjs/src/embed/hostEventClient/contracts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UIPassthroughEvent = exports.ApplicabilityLevel = void 0;
const types_1 = require("../../types");
/**
 * Levels at which a filter or parameter can be applied.
 */
var ApplicabilityLevel;
(function (ApplicabilityLevel) {
    ApplicabilityLevel["Liveboard"] = "LIVEBOARD";
    ApplicabilityLevel["Tab"] = "TAB";
    ApplicabilityLevel["Group"] = "GROUP";
})(ApplicabilityLevel = exports.ApplicabilityLevel || (exports.ApplicabilityLevel = {}));
var UIPassthroughEvent;
(function (UIPassthroughEvent) {
    UIPassthroughEvent["PinAnswerToLiveboard"] = "addVizToPinboard";
    UIPassthroughEvent["SaveAnswer"] = "saveAnswer";
    UIPassthroughEvent["GetDiscoverabilityStatus"] = "getDiscoverabilityStatus";
    UIPassthroughEvent["GetAvailableUIPassthroughs"] = "getAvailableUiPassthroughs";
    UIPassthroughEvent["GetAnswerConfig"] = "getAnswerPageConfig";
    UIPassthroughEvent["GetLiveboardConfig"] = "getPinboardPageConfig";
    UIPassthroughEvent["GetUnsavedAnswerTML"] = "getUnsavedAnswerTML";
    UIPassthroughEvent["UpdateFilters"] = "updateFilters";
    UIPassthroughEvent["Drilldown"] = "drillDown";
    UIPassthroughEvent["GetAnswerSession"] = "getAnswerSession";
    UIPassthroughEvent["GetFilters"] = "getFilters";
    UIPassthroughEvent["GetIframeUrl"] = "getIframeUrl";
    UIPassthroughEvent["GetParameters"] = "getParameters";
    UIPassthroughEvent["GetTML"] = "getTML";
    UIPassthroughEvent["GetTabs"] = "getTabs";
    UIPassthroughEvent["GetExportRequestForCurrentPinboard"] = "getExportRequestForCurrentPinboard";
})(UIPassthroughEvent = exports.UIPassthroughEvent || (exports.UIPassthroughEvent = {}));
