// @thoughtspot/visual-embed-sdk/cjs/src/embed/hostEventClient/utils.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("./utils");
const errors_1 = require("../../errors");
const types_1 = require("../../types");
const utils_2 = require("../../utils");
describe('hostEventClient utils', () => {
    // =========================
    // UpdateFilters Validation
    // =========================
    describe('isValidUpdateFiltersPayload', () => {
        it('returns false for undefined', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)(undefined)).toBe(false);
        });
        it('returns false for empty payload', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({})).toBe(false);
        });
        it('returns true for valid filter with column', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { column: 'region', oper: 'EQ', values: ['North'] },
            })).toBe(true);
        });
        it('returns true for valid filter with columnName', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { columnName: 'region', oper: 'EQ', values: ['North'] },
            })).toBe(true);
        });
        it('returns true for filter with operator instead of oper', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { column: 'region', operator: 'EQ', values: ['North'] },
            })).toBe(true);
        });
        it('returns true for valid filters array', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filters: [
                    { column: 'x', oper: 'IN', values: ['a', 'b'] },
                    { column: 'y', oper: 'EQ', values: ['c'] },
                ],
            })).toBe(true);
        });
        it('returns true for valid filters array with columnName', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filters: [
                    { columnName: 'x', oper: 'IN', values: ['a', 'b'] },
                    { columnName: 'y', oper: 'EQ', values: ['c'] },
                ],
            })).toBe(true);
        });
        it('returns false if one filter in filters array is invalid', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filters: [
                    { column: 'x', oper: 'EQ', values: ['a'] },
                    { column: 'y', values: ['b'] }, // invalid
                ],
            })).toBe(false);
        });
        it('returns false for filter with missing column/columnName', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { oper: 'EQ', values: ['a'] },
            })).toBe(false);
        });
        it('returns false for filter with missing operator', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { column: 'x', values: ['a'] },
            })).toBe(false);
        });
        it('returns false for filter with non-array values', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { column: 'x', oper: 'EQ', values: 'a' },
            })).toBe(false);
        });
        it('returns false for filter with non-string type', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { column: 'x', oper: 'EQ', values: ['a'], type: 123 },
            })).toBe(false);
        });
        it('returns true for filter with valid string type', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { column: 'x', oper: 'EQ', values: ['a'], type: 'STRING' },
            })).toBe(true);
        });
        it('returns false for empty filters array', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({ filters: [] })).toBe(false);
        });
        it('returns false if filters is not an array', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filters: 'invalid',
            })).toBe(false);
        });
        it('returns true if filter is valid even when filters is invalid', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { column: 'x', oper: 'EQ', values: ['a'] },
                filters: [{ column: 'y' }], // invalid
            })).toBe(true);
        });
        it('returns true for filter without applicability', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: { column: 'x', oper: 'EQ', values: ['a'] },
            })).toBe(true);
        });
        it('returns true for filter with valid applicability', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: {
                    column: 'x',
                    oper: 'EQ',
                    values: ['a'],
                    applicability: { level: 'TAB', targetId: 'tab-guid-1' },
                },
            })).toBe(true);
        });
        it('returns true for filters array with applicability on some filters', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filters: [
                    {
                        column: 'x',
                        oper: 'IN',
                        values: ['a', 'b'],
                        applicability: { level: 'TAB', targetId: 'tab-guid-1' },
                    },
                    { column: 'y', oper: 'EQ', values: ['c'] },
                ],
            })).toBe(true);
        });
        it('returns false for applicability missing targetId', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: {
                    column: 'x',
                    oper: 'EQ',
                    values: ['a'],
                    applicability: { level: 'TAB' },
                },
            })).toBe(false);
        });
        it('returns false for applicability with empty targetId', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: {
                    column: 'x',
                    oper: 'EQ',
                    values: ['a'],
                    applicability: { level: 'TAB', targetId: '' },
                },
            })).toBe(false);
        });
        it('returns false for null applicability', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: {
                    column: 'x',
                    oper: 'EQ',
                    values: ['a'],
                    applicability: null,
                },
            })).toBe(false);
        });
        it('returns false for non-object applicability', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: {
                    column: 'x',
                    oper: 'EQ',
                    values: ['a'],
                    applicability: 'TAB',
                },
            })).toBe(false);
        });
        it('returns true for LIVEBOARD level applicability without targetId', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: {
                    column: 'x',
                    oper: 'EQ',
                    values: ['a'],
                    applicability: { level: 'LIVEBOARD' },
                },
            })).toBe(true);
        });
        it('returns false for applicability missing level', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: {
                    column: 'x',
                    oper: 'EQ',
                    values: ['a'],
                    applicability: { targetId: 'tab-guid-1' },
                },
            })).toBe(false);
        });
        it('returns false for applicability with non-string level', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filter: {
                    column: 'x',
                    oper: 'EQ',
                    values: ['a'],
                    applicability: { level: 123, targetId: 'tab-guid-1' },
                },
            })).toBe(false);
        });
        it('returns false if one filter in filters array has invalid applicability', () => {
            expect((0, utils_1.isValidUpdateFiltersPayload)({
                filters: [
                    { column: 'x', oper: 'EQ', values: ['a'] },
                    {
                        column: 'y',
                        oper: 'EQ',
                        values: ['b'],
                        applicability: { level: 'TAB' }, // missing targetId
                    },
                ],
            })).toBe(false);
        });
    });
    // =========================
    // UpdateParameters Validation
    // =========================
    describe('isValidUpdateParametersPayload', () => {
        it('returns true for undefined payload', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)(undefined)).toBe(true);
        });
        it('returns true for non-array payload', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)({ name: 'p', value: 1 })).toBe(true);
        });
        it('returns true for parameters without applicability', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)([
                { name: 'p1', value: 1 },
                { name: 'p2', value: 'a' },
            ])).toBe(true);
        });
        it('returns true for null applicability', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)([
                { name: 'p', value: 1, applicability: null },
            ])).toBe(true);
        });
        it('returns true for valid TAB applicability with targetId', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)([
                { name: 'p', value: 1, applicability: { level: 'TAB', targetId: 'tab-guid-1' } },
            ])).toBe(true);
        });
        it('returns true for LIVEBOARD level applicability without targetId', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)([
                { name: 'p', value: 1, applicability: { level: 'LIVEBOARD' } },
            ])).toBe(true);
        });
        it('returns false for applicability missing targetId', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)([
                { name: 'p', value: 1, applicability: { level: 'TAB' } },
            ])).toBe(false);
        });
        it('returns false for applicability with empty targetId', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)([
                { name: 'p', value: 1, applicability: { level: 'GROUP', targetId: ' ' } },
            ])).toBe(false);
        });
        it('returns false for applicability with invalid level', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)([
                { name: 'p', value: 1, applicability: { level: 'VIZ', targetId: 'guid-1' } },
            ])).toBe(false);
        });
        it('returns false for non-object applicability', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)([
                { name: 'p', value: 1, applicability: 'TAB' },
            ])).toBe(false);
        });
        it('returns false if one parameter has invalid applicability', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)([
                { name: 'p1', value: 1 },
                { name: 'p2', value: 2, applicability: { level: 'GROUP' } }, // missing targetId
            ])).toBe(false);
        });
        it('returns true for non-object entries', () => {
            expect((0, utils_1.isValidUpdateParametersPayload)(['p', 1, null])).toBe(true);
        });
    });
    // =========================
    // DrillDown Validation
    // =========================
    describe('isValidDrillDownPayload', () => {
        it('returns false for undefined', () => {
            expect((0, utils_1.isValidDrillDownPayload)(undefined)).toBe(false);
        });
        it('returns false for empty payload', () => {
            expect((0, utils_1.isValidDrillDownPayload)({})).toBe(false);
        });
        it('returns false for empty points object', () => {
            expect((0, utils_1.isValidDrillDownPayload)({ points: {} })).toBe(false);
        });
        it('returns false for non-object points', () => {
            expect((0, utils_1.isValidDrillDownPayload)({ points: 'invalid' })).toBe(false);
        });
        it('returns true for clickedPoint', () => {
            expect((0, utils_1.isValidDrillDownPayload)({
                points: { clickedPoint: 'point-1' },
            })).toBe(true);
        });
        it('returns true for selectedPoints', () => {
            expect((0, utils_1.isValidDrillDownPayload)({
                points: { selectedPoints: ['p1', 'p2'] },
            })).toBe(true);
        });
        it('returns true for both clickedPoint and selectedPoints', () => {
            expect((0, utils_1.isValidDrillDownPayload)({
                points: { clickedPoint: 'p1', selectedPoints: ['p2'] },
            })).toBe(true);
        });
        it('returns false for selectedPoints empty array', () => {
            expect((0, utils_1.isValidDrillDownPayload)({
                points: { selectedPoints: [] },
            })).toBe(false);
        });
        it('returns false if clickedPoint is null', () => {
            expect((0, utils_1.isValidDrillDownPayload)({
                points: { clickedPoint: null },
            })).toBe(false);
        });
        it('returns false if selectedPoints is not an array', () => {
            expect((0, utils_1.isValidDrillDownPayload)({
                points: { selectedPoints: 'invalid' },
            })).toBe(false);
        });
        it('returns false if both clickedPoint and selectedPoints are invalid', () => {
            expect((0, utils_1.isValidDrillDownPayload)({
                points: { clickedPoint: null, selectedPoints: [] },
            })).toBe(false);
        });
    });
    // =========================
    // Error Handling
    // =========================
    describe('createValidationError', () => {
        it('throws with message', () => {
            expect(() => (0, utils_1.createValidationError)('test error')).toThrow('test error');
        });
        it('throws Error instance with metadata', () => {
            try {
                (0, utils_1.createValidationError)('custom msg');
            }
            catch (err) {
                expect(err).toBeInstanceOf(Error);
                expect(err.isValidationError).toBe(true);
                expect(err.embedErrorDetails).toBeDefined();
                expect(err.embedErrorDetails).toMatchObject({
                    type: types_1.EmbedEvent.Error,
                    data: {
                        errorType: 'VALIDATION_ERROR',
                        message: 'custom msg',
                        code: 'HOST_EVENT_VALIDATION',
                        error: 'custom msg',
                    },
                    status: utils_2.embedEventStatus.END,
                });
            }
        });
    });
    describe('throwUpdateFiltersValidationError', () => {
        it('throws with UPDATEFILTERS_INVALID_PAYLOAD message', () => {
            expect(() => (0, utils_1.throwUpdateFiltersValidationError)())
                .toThrow(errors_1.ERROR_MESSAGE.UPDATEFILTERS_INVALID_PAYLOAD);
        });
    });
    describe('throwUpdateParametersValidationError', () => {
        it('throws with UPDATEPARAMETERS_INVALID_PAYLOAD message', () => {
            expect(() => (0, utils_1.throwUpdateParametersValidationError)())
                .toThrow(errors_1.ERROR_MESSAGE.UPDATEPARAMETERS_INVALID_PAYLOAD);
        });
    });
    describe('throwDrillDownValidationError', () => {
        it('throws with DRILLDOWN_INVALID_PAYLOAD message', () => {
            expect(() => (0, utils_1.throwDrillDownValidationError)())
                .toThrow(errors_1.ERROR_MESSAGE.DRILLDOWN_INVALID_PAYLOAD);
        });
    });
});
