// @thoughtspot/visual-embed-sdk/cjs/src/embed/hostEventClient/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.throwUpdateParametersValidationError = exports.throwDrillDownValidationError = exports.throwUpdateFiltersValidationError = exports.createValidationError = exports.isValidDrillDownPayload = exports.isValidUpdateParametersPayload = exports.isValidUpdateFiltersPayload = void 0;
const tslib_1 = require("tslib");
const isNil_1 = tslib_1.__importDefault(require("lodash/isNil"));
const isPlainObject_1 = tslib_1.__importDefault(require("lodash/isPlainObject"));
const isString_1 = tslib_1.__importDefault(require("lodash/isString"));
const isUndefined_1 = tslib_1.__importDefault(require("lodash/isUndefined"));
const types_1 = require("../../types");
const errors_1 = require("../../errors");
const contracts_1 = require("./contracts");
const utils_1 = require("../../utils");
const isValidApplicability = (a) => {
    if ((0, isUndefined_1.default)(a))
        return true;
    // targetId is not required at LIVEBOARD level, since the filter applies to the whole Liveboard
    return (0, isPlainObject_1.default)(a)
        && Object.values(contracts_1.ApplicabilityLevel).includes(a.level)
        && (a.level === contracts_1.ApplicabilityLevel.Liveboard || ((0, isString_1.default)(a.targetId) && a.targetId.trim().length > 0));
};
function isValidUpdateFiltersPayload(payload) {
    if (!payload)
        return false;
    const isValidFilter = (f) => {
        const hasColumn = typeof f.column === 'string' || typeof f.columnName === 'string';
        const hasOperator = typeof f.oper === 'string' || typeof f.operator === 'string';
        const hasValues = Array.isArray(f.values);
        const validType = !f.type || typeof f.type === 'string';
        return hasColumn && hasOperator && hasValues && validType && isValidApplicability(f.applicability);
    };
    const hasValidFilter = payload.filter && isValidFilter(payload.filter);
    const hasValidFilters = Array.isArray(payload.filters) && payload.filters.length > 0 && payload.filters.every(isValidFilter);
    return !!(hasValidFilter || hasValidFilters);
}
exports.isValidUpdateFiltersPayload = isValidUpdateFiltersPayload;
function isValidUpdateParametersPayload(payload) {
    // Only validates the applicability of each parameter (null treated as absent); the rest is forwarded as-is for backward compatibility.
    if (!Array.isArray(payload))
        return true;
    return payload.every((p) => {
        if (!(0, isPlainObject_1.default)(p))
            return true;
        const { applicability } = p;
        return (0, isNil_1.default)(applicability) || isValidApplicability(applicability);
    });
}
exports.isValidUpdateParametersPayload = isValidUpdateParametersPayload;
function isValidDrillDownPayload(payload) {
    if (!payload)
        return false;
    const points = payload.points;
    if (!points || typeof points !== 'object')
        return false;
    const hasClickedPoint = 'clickedPoint' in points && points.clickedPoint != null;
    const hasSelectedPoints = Array.isArray(points.selectedPoints) && points.selectedPoints.length > 0;
    return hasClickedPoint || hasSelectedPoints;
}
exports.isValidDrillDownPayload = isValidDrillDownPayload;
function createValidationError(message) {
    const err = new Error(message);
    err.isValidationError = true;
    err.embedErrorDetails = {
        type: types_1.EmbedEvent.Error,
        data: {
            errorType: types_1.ErrorDetailsTypes.VALIDATION_ERROR,
            message,
            code: types_1.EmbedErrorCodes.HOST_EVENT_VALIDATION,
            error: message
        },
        status: utils_1.embedEventStatus.END
    };
    throw err;
}
exports.createValidationError = createValidationError;
function throwUpdateFiltersValidationError() {
    createValidationError(errors_1.ERROR_MESSAGE.UPDATEFILTERS_INVALID_PAYLOAD);
}
exports.throwUpdateFiltersValidationError = throwUpdateFiltersValidationError;
function throwDrillDownValidationError() {
    createValidationError(errors_1.ERROR_MESSAGE.DRILLDOWN_INVALID_PAYLOAD);
}
exports.throwDrillDownValidationError = throwDrillDownValidationError;
function throwUpdateParametersValidationError() {
    createValidationError(errors_1.ERROR_MESSAGE.UPDATEPARAMETERS_INVALID_PAYLOAD);
}
exports.throwUpdateParametersValidationError = throwUpdateParametersValidationError;
